From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Fri, 6 Jan 2023 17:11:31 +0900
Subject: [PATCH] Publish Packages


diff --git a/build.gradle.kts b/build.gradle.kts
index a85f1db096555a6077dd77cbd7572442d3adc848..73cf94c26512ac622948e8ec5422beed3ac30bae 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -153,3 +153,23 @@ tasks.check {
     dependsOn(scanJar)
 }
 // Paper end
+
+// Andromeda start
+publishing {
+    repositories {
+        maven {
+            name = "githubPackage"
+            url = uri("https://maven.pkg.github.com/TeamEarendel/Andromeda")
+
+            credentials.username = System.getenv("GITHUB_USERNAME")
+            credentials.password = System.getenv("GITHUB_TOKEN")
+        }
+
+        publications {
+            register<MavenPublication>("gpr") {
+                from(components["java"])
+            }
+        }
+    }
+}
+// Andromeda end
