From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sun, 15 Jan 2023 16:03:20 +0900
Subject: [PATCH] Mining speed tweaks


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index cfce3d279d5f6961e7284753d7e1b8f7bac0b33e..2e90e8afc02ca2482ae52c80a5b7a0bb0ff9ac05 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3143,5 +3143,19 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param component Name of the server mod to display to the Player
      */
     void setModName(@NotNull Component component);
+
+    /**
+     * Gets the mining amplifier for the player.
+     *
+     * @return Mining amplifier for the player
+     */
+    float getMiningAmplifier();
+
+    /**
+     * Sets the mining amplifier of the player.
+     *
+     * @param amplifier mining amplifier
+     */
+    void setMiningAmplifier(float amplifier);
     // Andromeda end
 }
