From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 18:22:40 +0900
Subject: [PATCH] Add server mod name API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index d5ad02c1fa2b840289d316400952aaebb8c5dad1..cfce3d279d5f6961e7284753d7e1b8f7bac0b33e 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3135,4 +3135,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     void sendDeathScreen(@NotNull Component message, @Nullable Entity killer);
     // Purpur end
+
+    // Andromeda start
+    /**
+     * Sets the name of the server mod name displayed to the player.
+     *
+     * @param component Name of the server mod to display to the Player
+     */
+    void setModName(@NotNull Component component);
+    // Andromeda end
 }
