From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 18:22:40 +0900
Subject: [PATCH] Add server mod name API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index f65a860d09d9f1d5045b2af63a80109e7c1c4ddf..ff6a06ef41827d076ad0b89bb5c8c090ac2dc443 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3063,4 +3063,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     void sendDeathScreen(@NotNull Component message, @Nullable Entity killer);
     // Purpur end
+
+    // Andromeda start
+    /**
+     * Sets the name of the server mod name displayed to the player.
+     *
+     * @param component Name of the server mod to display to the Player
+     */
+    void setModName(@NotNull Component component);
+    // Andromeda end
 }
