From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Mon, 16 Jan 2023 20:51:57 +0900
Subject: [PATCH] Port PaperMC/Paper#8790


diff --git a/src/main/java/com/destroystokyo/paper/event/entity/PreCreatureSpawnEvent.java b/src/main/java/com/destroystokyo/paper/event/entity/PreCreatureSpawnEvent.java
index 3ad231aa3206c8cfd5ec995249584cebab5d11f3..eb6e8b55b2b1a6cf6b3ca76e7fab98ac19de4a8b 100644
--- a/src/main/java/com/destroystokyo/paper/event/entity/PreCreatureSpawnEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/entity/PreCreatureSpawnEvent.java
@@ -26,7 +26,7 @@ public class PreCreatureSpawnEvent extends Event implements Cancellable {
     private boolean shouldAbortSpawn;
 
     public PreCreatureSpawnEvent(@NotNull Location location, @NotNull EntityType type, @NotNull CreatureSpawnEvent.SpawnReason reason) {
-        this.location = Preconditions.checkNotNull(location, "Location may not be null").clone();
+        this.location = Preconditions.checkNotNull(location, "Location may not be null"); // Andromeda - Port PaperMC/Paper#8790
         this.type = Preconditions.checkNotNull(type, "Type may not be null");
         this.reason = Preconditions.checkNotNull(reason, "Reason may not be null");
     }
