From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Thu, 22 Dec 2022 20:32:39 +0900
Subject: [PATCH] Add log that Spark is downloaded and installed


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index a3ce01cdd3d4c97a0b4b1a3881f739ed70b2a7e9..a1549d49135a2b5b5391759fcb3a0bf29e3ea5f6 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -473,6 +473,7 @@ public final class CraftServer implements Server {
                     }
 
                     extraJars.add(file);
+                    getLogger().info("Purpur: Spark has been downloaded and installed. If you really don't want to use Spark, add \"-DPurpur.IReallyDontWantSpart=true\" to your server startup flags, IN FRONT OF the \"-jar\""); // Andromeda
                 } catch (Exception e) {
                     getLogger().severe("Purpur: Failed to download and install spark plugin");
                     e.printStackTrace();
