From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Thu, 22 Dec 2022 20:32:39 +0900
Subject: [PATCH] Add log that Spark is downloaded and installed


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index a7ec7645d79018a542d2d6b59a95789f09291701..68aa8b2de392e2527239f0c515a55abc2d93e198 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -471,6 +471,7 @@ public final class CraftServer implements Server {
                     }
 
                     extraJars.add(file);
+                    getLogger().info("Purpur: Spark has been downloaded and installed. If you really don't want to use Spark, add \"-DPurpur.IReallyDontWantSpart=true\" to your server startup flags, BETWEEN the \"-jar\" and <server file name>"); // Andromeda
                 } catch (Exception e) {
                     getLogger().severe("Purpur: Failed to download and install spark plugin");
                     e.printStackTrace();
