From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Thu, 22 Dec 2022 20:52:01 +0900
Subject: [PATCH] Add option to configure portal size


diff --git a/src/main/java/net/minecraft/world/level/portal/PortalShape.java b/src/main/java/net/minecraft/world/level/portal/PortalShape.java
index e7554ec800f321e4e34c926c53f2375a8c3aa979..c80cbe3728df6e66eb4fa5e0dd65bf09b555b62f 100644
--- a/src/main/java/net/minecraft/world/level/portal/PortalShape.java
+++ b/src/main/java/net/minecraft/world/level/portal/PortalShape.java
@@ -100,7 +100,7 @@ public class PortalShape {
     private int calculateWidth() {
         int i = this.getDistanceUntilEdgeAboveFrame(this.bottomLeft, this.rightDir);
 
-        return i >= 2 && i <= 21 ? i : 0;
+        return i >= this.level.getMinecraftWorld().andromedaLevelConfiguration().structure.portal.netherPortal.minimumWidth() && (this.level.getMinecraftWorld().andromedaLevelConfiguration().structure.portal.netherPortal.maximumWidth() == -1 || i <= this.level.getMinecraftWorld().andromedaLevelConfiguration().structure.portal.netherPortal.maximumWidth()) ? i : 0; // Andromeda
     }
 
     private int getDistanceUntilEdgeAboveFrame(BlockPos pos, Direction direction) {
@@ -133,7 +133,7 @@ public class PortalShape {
         BlockPos.MutableBlockPos blockposition_mutableblockposition = new BlockPos.MutableBlockPos();
         int i = this.getDistanceUntilTop(blockposition_mutableblockposition);
 
-        return i >= 3 && i <= 21 && this.hasTopFrame(blockposition_mutableblockposition, i) ? i : 0;
+        return i >= this.level.getMinecraftWorld().andromedaLevelConfiguration().structure.portal.netherPortal.minimumHeight() && (this.level.getMinecraftWorld().andromedaLevelConfiguration().structure.portal.netherPortal.maximunHeight() == -1 || i <= this.level.getMinecraftWorld().andromedaLevelConfiguration().structure.portal.netherPortal.maximunHeight()) && this.hasTopFrame(blockposition_mutableblockposition, i) ? i : 0; // Andromeda
     }
 
     private boolean hasTopFrame(BlockPos.MutableBlockPos pos, int height) {
@@ -187,7 +187,15 @@ public class PortalShape {
     }
 
     public boolean isValid() {
-        return this.bottomLeft != null && this.width >= 2 && this.width <= 21 && this.height >= 3 && this.height <= 21;
+        // Andromeda start
+        boolean i = this.width >= this.level.getMinecraftWorld().andromedaLevelConfiguration().structure.portal.netherPortal.minimumWidth()
+                && (this.level.getMinecraftWorld().andromedaLevelConfiguration().structure.portal.netherPortal.maximumWidth() == -1
+                || this.width <= this.level.getMinecraftWorld().andromedaLevelConfiguration().structure.portal.netherPortal.maximumWidth());
+        boolean j = this.height >= this.level.getMinecraftWorld().andromedaLevelConfiguration().structure.portal.netherPortal.minimumHeight()
+                && (this.level.getMinecraftWorld().andromedaLevelConfiguration().structure.portal.netherPortal.maximunHeight() == -1
+                || this.width <= this.level.getMinecraftWorld().andromedaLevelConfiguration().structure.portal.netherPortal.maximunHeight());
+        return this.bottomLeft != null && i && j;
+        // Andromeda end
     }
 
     // CraftBukkit start - return boolean
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index ec0e557a099abd76789684c515fe2c5c4128288e..4a57c3f65a6d279c3bc9ebe5401dee5ec84ed510 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -34,4 +34,24 @@ public class LevelConfigurations extends ConfigurationPart {
 
     }
 
+    public Structure structure;
+    public class Structure extends ConfigurationPart {
+
+        public Portal portal;
+        public class Portal extends ConfigurationPart {
+
+            public NetherPortal netherPortal;
+            public class NetherPortal extends ConfigurationPart {
+
+                public int minimumWidth = 2; public int minimumWidth() { return Math.max(this.minimumWidth, 1); }
+                public int minimumHeight = 3; public int minimumHeight() { return Math.max(this.minimumHeight, 2); }
+                public int maximumWidth = 21; public int maximumWidth() { return Math.max(this.maximumWidth, this.minimumWidth); }
+                public int maximunHeight = 21; public int maximunHeight() { return Math.max(this.maximunHeight, this.minimumHeight); }
+
+            }
+
+        }
+
+    }
+
 }
