From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 24 Dec 2022 19:48:15 +0900
Subject: [PATCH] Enchantment Tweaks


diff --git a/src/main/java/net/minecraft/world/item/enchantment/DamageEnchantment.java b/src/main/java/net/minecraft/world/item/enchantment/DamageEnchantment.java
index d6c354856c71d9629828e2ddbe316306bfaf6ea6..e25aa510aa03bc32ea8da6b0e5308769dbadebd7 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/DamageEnchantment.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/DamageEnchantment.java
@@ -47,11 +47,13 @@ public class DamageEnchantment extends Enchantment {
 
     @Override
     public boolean checkCompatibility(Enchantment other) {
+        if (team.earendel.andromeda.configurations.GlobalConfiguration.get().enchantments.mixableWeaponEnchantments) return true; // Andromeda - Enchantment Tweaks
         return !(other instanceof DamageEnchantment);
     }
 
     @Override
     public boolean canEnchant(ItemStack stack) {
+        if (team.earendel.andromeda.configurations.GlobalConfiguration.get().enchantments.trident.acceptWeaponEnchantments && stack.getItem() instanceof net.minecraft.world.item.TridentItem) return true; // Andromeda - Enchantment Tweaks
         return stack.getItem() instanceof AxeItem ? true : super.canEnchant(stack);
     }
 
diff --git a/src/main/java/net/minecraft/world/item/enchantment/Enchantment.java b/src/main/java/net/minecraft/world/item/enchantment/Enchantment.java
index 2381f72a5e980b8659f6253054996bbcad2844fc..19be3d4782a5e3857c5a20b463bd94f8bc13d56d 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/Enchantment.java
@@ -110,6 +110,18 @@ public abstract class Enchantment {
     }
 
     public boolean canEnchant(ItemStack stack) {
+        // Andromeda start
+        if (this instanceof FireAspectEnchantment) {
+            if (stack.getItem() instanceof net.minecraft.world.item.TridentItem) return team.earendel.andromeda.configurations.GlobalConfiguration.get().enchantments.trident.acceptFireAspect;
+            else if (stack.getItem() instanceof net.minecraft.world.item.AxeItem) return team.earendel.andromeda.configurations.GlobalConfiguration.get().enchantments.axe.acceptFireAspect;
+        } else if (this instanceof KnockbackEnchantment) {
+            if (stack.getItem() instanceof net.minecraft.world.item.TridentItem) return team.earendel.andromeda.configurations.GlobalConfiguration.get().enchantments.trident.acceptKnockback;
+            else if (stack.getItem() instanceof net.minecraft.world.item.AxeItem) return team.earendel.andromeda.configurations.GlobalConfiguration.get().enchantments.axe.acceptKnockback;
+        } else if (this instanceof LootBonusEnchantment) {
+            if (stack.getItem() instanceof net.minecraft.world.item.TridentItem) return team.earendel.andromeda.configurations.GlobalConfiguration.get().enchantments.trident.acceptLooting;
+            else if (stack.getItem() instanceof net.minecraft.world.item.AxeItem) return team.earendel.andromeda.configurations.GlobalConfiguration.get().enchantments.axe.acceptFireAspect;
+        }
+        // Andromeda end
         return this.category.canEnchant(stack.getItem());
     }
 
diff --git a/src/main/java/net/minecraft/world/item/enchantment/ProtectionEnchantment.java b/src/main/java/net/minecraft/world/item/enchantment/ProtectionEnchantment.java
index a38194accb2d0ba204ab2f81b4be6031842faf9c..963ac085c994db466f1d05ecf295d0b6a67adab0 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/ProtectionEnchantment.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/ProtectionEnchantment.java
@@ -51,6 +51,7 @@ public class ProtectionEnchantment extends Enchantment {
             if (this.type == protectionEnchantment.type) {
                 return false;
             } else {
+                if (team.earendel.andromeda.configurations.GlobalConfiguration.get().enchantments.mixableProtectionEnchantments) return true; // Andromeda - Enchantment Tweaks
                 return this.type == ProtectionEnchantment.Type.FALL || protectionEnchantment.type == ProtectionEnchantment.Type.FALL;
             }
         } else {
diff --git a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
index 5fae009c978110663281234286ba05200587fbcf..918f5d9649971e2d595e598afad526758d693f1b 100644
--- a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
+++ b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
@@ -29,4 +29,25 @@ public class GlobalConfiguration extends ConfigurationPart {
     public class ConsoleLogs extends ConfigurationPart {
         public boolean offlineWarnings = true;
     }
+
+    public Enchantments enchantments;
+    public class Enchantments extends ConfigurationPart {
+        public boolean mixableWeaponEnchantments = false;
+        public boolean mixableProtectionEnchantments = false;
+
+        public Trident trident;
+        public class Trident extends ConfigurationPart {
+            public boolean acceptWeaponEnchantments = false;
+            public boolean acceptFireAspect = false;
+            public boolean acceptKnockback = false;
+            public boolean acceptLooting = false;
+        }
+
+        public Axe axe;
+        public class Axe extends ConfigurationPart {
+            public boolean acceptFireAspect = false;
+            public boolean acceptKnockback = false;
+            public boolean acceptLooting = false;
+        }
+    }
 }
