From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 24 Dec 2022 22:36:54 +0900
Subject: [PATCH] fixup! Rebrand


diff --git a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
index 9105be6d4000b6ba1ea89be8556e9a7be8f88523..abd9b2bb9b2079b1eef20e3b1ab1c5c24cc91e65 100644
--- a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
+++ b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
@@ -66,6 +66,9 @@ public class PaperVersionFetcher implements VersionFetcher {
             versionInfo = versionInfo.replace("\"", "");
             distance = fetchDistanceFromGitHub(repo, branch, versionInfo);
         }
+         */ // Andromeda - Disable CI Checking
+        versionInfo = versionInfo.replace("\"", ""); // Andromeda
+        distance = fetchDistanceFromGitHub(repo, branch, versionInfo); // Andromeda
 
         switch (distance) {
             case -1:
@@ -82,9 +85,6 @@ public class PaperVersionFetcher implements VersionFetcher {
                                         .hoverEvent(Component.text("Click to open", NamedTextColor.WHITE))
                                         .clickEvent(ClickEvent.openUrl(DOWNLOAD_PAGE))));
         }
-         */ // Andromeda - Disalbe CI Checking
-         versionInfo = versionInfo.replace("\"", ""); // Andromeda
-         distance = fetchDistanceFromGitHub(repo, branch, versionInfo); // Andromeda
     }
 
     /* // Andromeda - Disalbe CI Checking
