From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Tue, 27 Dec 2022 21:43:14 +0900
Subject: [PATCH] Add armor limit config


diff --git a/src/main/java/net/minecraft/world/damagesource/CombatRules.java b/src/main/java/net/minecraft/world/damagesource/CombatRules.java
index aa2331c6df4e79d4bb0add071a0b11d2a3a08b88..1e3c1f781de1e52493a5e391791710af147b4076 100644
--- a/src/main/java/net/minecraft/world/damagesource/CombatRules.java
+++ b/src/main/java/net/minecraft/world/damagesource/CombatRules.java
@@ -11,12 +11,12 @@ public class CombatRules {
 
     public static float getDamageAfterAbsorb(float damage, float armor, float armorToughness) {
         float f = 2.0F + armorToughness / 4.0F;
-        float g = Mth.clamp(armor - damage / f, armor * 0.2F, org.purpurmc.purpur.PurpurConfig.limitArmor ? 20F : Float.MAX_VALUE); // Purpur
+        float g = Mth.clamp(armor - damage / f, armor * 0.2F, org.purpurmc.purpur.PurpurConfig.limitArmor ? 20F : team.earendel.andromeda.configurations.GlobalConfiguration.get().attribute.armor.maximum); // Purpur // Andromeda
         return damage * (1.0F - g / 25.0F);
     }
 
     public static float getDamageAfterMagicAbsorb(float damageDealt, float protection) {
-        float f = Mth.clamp(protection, 0.0F, org.purpurmc.purpur.PurpurConfig.limitArmor ? 20F : Float.MAX_VALUE); // Purpur
+        float f = Mth.clamp(protection, 0.0F, org.purpurmc.purpur.PurpurConfig.limitArmor ? 20F : team.earendel.andromeda.configurations.GlobalConfiguration.get().attribute.armor.maximum); // Purpur // Andromeda
         return damageDealt * (1.0F - f / 25.0F);
     }
 }
diff --git a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
index 918f5d9649971e2d595e598afad526758d693f1b..bba22602522394da88d11f19e14a2dd7d45a0ec1 100644
--- a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
+++ b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
@@ -50,4 +50,14 @@ public class GlobalConfiguration extends ConfigurationPart {
             public boolean acceptLooting = false;
         }
     }
+
+    public Attribute attribute;
+    public class Attribute extends ConfigurationPart {
+
+        public Armor armor;
+        public class Armor extends ConfigurationPart {
+            public float maximum = 20.0F;
+        }
+
+    }
 }
