From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Fri, 6 Jan 2023 14:48:39 +0900
Subject: [PATCH] Add configuration to disable moved to quickly check for
 spectator


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index a8e205af936efe0909c94575e3b7f8ca50cead2f..35648444cb2044a18dfb5637ebbdd4c38d5d3399 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1531,7 +1531,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
                             if (!this.player.isChangingDimension() && (!this.player.getLevel().getGameRules().getBoolean(GameRules.RULE_DISABLE_ELYTRA_MOVEMENT_CHECK) || !this.player.isFallFlying())) {
                                 float f2 = this.player.isFallFlying() ? 300.0F : 100.0F;
 
-                                if (d11 - d10 > Math.max(f2, Math.pow((double) (org.spigotmc.SpigotConfig.movedTooQuicklyMultiplier * (float) i * speed), 2)) && !this.isSingleplayerOwner()) {
+                                if (d11 - d10 > Math.max(f2, Math.pow((double) (org.spigotmc.SpigotConfig.movedTooQuicklyMultiplier * (float) i * speed), 2)) && !this.isSingleplayerOwner() && (this.player.gameMode.getGameModeForPlayer() != GameType.SPECTATOR && this.player.getLevel().andromedaLevelConfiguration().player.checkSpectatorMovedToQuickly)) { // Andromeda
                                 // CraftBukkit end
                                     ServerGamePacketListenerImpl.LOGGER.warn("{} moved too quickly! {},{},{}", new Object[]{this.player.getName().getString(), d7, d8, d9});
                                     this.teleport(this.player.getX(), this.player.getY(), this.player.getZ(), this.player.getYRot(), this.player.getXRot());
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 481dd6d0377988b17a1bef1c51d103f961c5ea87..8265845110fa9472b5a528d1075fbcab14e6ae1c 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -73,4 +73,11 @@ public class LevelConfigurations extends ConfigurationPart {
 
     }
 
+    public Player player;
+    public class Player extends ConfigurationPart {
+
+        public boolean checkSpectatorMovedToQuickly = true;
+
+    }
+
 }
