From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Fri, 6 Jan 2023 17:37:48 +0900
Subject: [PATCH] Port Very Many Players


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 0ec4cb1c721b85b25ba0f4eac48e8d13113441f0..8241dc78f9aaa085657bd9e44cf54b1b6bde71fe 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -399,6 +399,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
     public boolean collidingWithWorldBorder; // Paper
     public @Nullable Boolean immuneToFire = null; // Purpur - Fire immune API
+    private boolean boundingBoxChanged = false; // Andromeda - Port VMP
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -1036,6 +1037,12 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     // Paper end - detailed watchdog information
 
     public void move(MoverType movementType, Vec3 movement) {
+        // Andromeda start - Port VMP
+        if (!boundingBoxChanged && movement.equals(Vec3.ZERO)) {
+            boundingBoxChanged = false;
+            return;
+        }
+        // Andromeda end
         // Paper start - detailed watchdog information
         io.papermc.paper.util.TickThread.ensureTickThread("Cannot move an entity off-main");
         synchronized (this.posLock) {
@@ -3847,6 +3854,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     }
 
     public final void setBoundingBox(AABB boundingBox) {
+        if (!this.bb.equals(boundingBox)) boundingBoxChanged = true; // Andromeda - Port VMP
         // CraftBukkit start - block invalid bounding boxes
         double minX = boundingBox.minX,
                 minY = boundingBox.minY,
diff --git a/src/main/java/team/earendel/andromeda/configurations/AndromedaWorldConfiguration.java b/src/main/java/team/earendel/andromeda/configurations/AndromedaWorldConfiguration.java
index 5e6c9f8034280bd2ac9c8fb98a68a976646ed709..4949a69fd8098d5187bcc03789a8555ff23f03ae 100644
--- a/src/main/java/team/earendel/andromeda/configurations/AndromedaWorldConfiguration.java
+++ b/src/main/java/team/earendel/andromeda/configurations/AndromedaWorldConfiguration.java
@@ -117,4 +117,9 @@ public class AndromedaWorldConfiguration {
     private void spectatorSettings() {
         checkSpectatorMovedToQuickly = getBoolean("spectator.check-moved-to-quickly", checkSpectatorMovedToQuickly);
     }
+
+    public boolean skipEntityMoveIfNoMovement = true;
+    private void veryManyPlayers() {
+        skipEntityMoveIfNoMovement = getBoolean("ports.very-many-players.skip-entity-move-if-no-movement", skipEntityMoveIfNoMovement);
+    }
 }
