From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 17:57:05 +0900
Subject: [PATCH] A Cursed Option


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index d85cd2019904c6cc419d708ea4eba8188dcf64f7..a0b0544594643c72643cbbc3c41f75873c25275e 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1334,6 +1334,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     // CraftBukkit end
 
     protected void waitUntilNextTick() {
+        if (team.earendel.andromeda.configurations.GlobalConfiguration.get().misc.unsupportedSettings.cursed.disableTPSLimit) return; // Andromeda - CURSED
         //this.executeAll(); // Paper - move this into the tick method for timings
         this.managedBlock(() -> {
             return !this.canSleepForTickNoOversleep(); // Paper - move oversleep into full server tick
diff --git a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
index 4c63c5a110ff315030ab9df8f224cd21008a5337..7f657878cb2e8a6007dd7471ec6eb669b05d5a7e 100644
--- a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
+++ b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
@@ -125,6 +125,18 @@ public class GlobalConfiguration extends ConfigurationPart {
 
         public boolean useFasterRandom = true;
 
+        public UnsupportedSettings unsupportedSettings;
+        public class UnsupportedSettings extends ConfigurationPart {
+
+            public Cursed cursed;
+            public class Cursed extends ConfigurationPart {
+
+                public boolean disableTPSLimit = false;
+
+            }
+
+        }
+
     }
 
     public NeverTickAlone neverTickAlone;
