From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 18:17:45 +0900
Subject: [PATCH] Add configuration to disable advancement reward


diff --git a/src/main/java/net/minecraft/advancements/AdvancementRewards.java b/src/main/java/net/minecraft/advancements/AdvancementRewards.java
index ca24b2373dfdcd887eacfdd5c4299e6dd61f63f7..e3921fac9a16a0da17ab79508c3dc6a2b8c0e881 100644
--- a/src/main/java/net/minecraft/advancements/AdvancementRewards.java
+++ b/src/main/java/net/minecraft/advancements/AdvancementRewards.java
@@ -42,7 +42,7 @@ public class AdvancementRewards {
     }
 
     public void grant(ServerPlayer player) {
-        player.giveExperiencePoints(this.experience);
+        if (player.getLevel().andromedaLevelConfiguration().player.giveAdvancementsReward) player.giveExperiencePoints(this.experience); // Andromeda
         LootContext lootContext = (new LootContext.Builder(player.getLevel())).withParameter(LootContextParams.THIS_ENTITY, player).withParameter(LootContextParams.ORIGIN, player.position()).withRandom(player.getRandom()).create(LootContextParamSets.ADVANCEMENT_REWARD);
         boolean bl = false;
 
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 710d04666edc734f62581211a0c99534afa40efd..9e7fd624ce046129e63994f2737cca2079f97be6 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -77,6 +77,7 @@ public class LevelConfigurations extends ConfigurationPart {
     public class Player extends ConfigurationPart {
 
         public boolean checkSpectatorMovedToQuickly = true;
+        public boolean giveAdvancementsReward = true;
 
     }
 
