From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 18:42:00 +0900
Subject: [PATCH] Port PurpurMC/Purpur#1234


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 0be544660aa5b2131cd8b985e14f1a2fda25e3b2..75e692c8b776bacb96791b083641a72cc734eb42 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -141,6 +141,23 @@ public class Explosion {
             return;
         }
         // CraftBukkit end
+
+        // Andromeda start - Port PurpurMC/Purpur#1234
+        Location loc = new Location(this.level.getWorld(), this.x, this.y, this.z);
+        float yield = this.blockInteraction == BlockInteraction.DESTROY_WITH_DECAY ? 1.0F / this.radius : 1.0F;
+        if (this.source != null) {
+            if (!(new org.purpurmc.purpur.event.entity.PreEntityExplodeEvent(this.source.getBukkitEntity(), loc, yield).callEvent())) {
+                this.wasCanceled = true;
+                return;
+            }
+        } else {
+            if (!(new org.purpurmc.purpur.event.PreBlockExplodeEvent(loc.getBlock(), yield).callEvent())) {
+                this.wasCanceled = true;
+                return;
+            }
+        }
+        // Andromeda end
+
         this.level.gameEvent(this.source, GameEvent.EXPLODE, new Vec3(this.x, this.y, this.z));
         Set<BlockPos> set = Sets.newHashSet();
         boolean flag = true;
