From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 19:38:53 +0900
Subject: [PATCH] Port PaperMC/Paper#8754


diff --git a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
index 01d8f1c985223eec83beb4a84a710d1bf8d70849..e39f90f080970db0a35c244e2faa167742953df0 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -87,11 +87,19 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
                         result = blockCookEvent.getResult();
                         itemstack1 = CraftItemStack.asNMSCopy(result);
                         // CraftBukkit end
-                    // Paper start
-                    net.minecraft.world.entity.item.ItemEntity droppedItem = new net.minecraft.world.entity.item.ItemEntity(world, pos.getX() + 0.5D, pos.getY() + 0.5D, pos.getZ() + 0.5D, itemstack1.split(world.random.nextInt(21) + 10));
-                    droppedItem.setDeltaMovement(world.random.nextGaussian() * 0.05D, world.random.nextGaussian() * 0.05D + 0.2D, world.random.nextGaussian() * 0.05D);
-                    world.addFreshEntity(droppedItem);
-                    // Paper end
+                        // Andromeda start - Port PaperMC/Paper#8754
+                        boolean flag1 = world.andromedaLevelConfiguration().paperPullRequestsPort.fixCampfireRecipesNotAlwaysOutputtingFullResult;
+                        while (!itemstack1.isEmpty() || !flag1) {
+                            // Paper start
+                            net.minecraft.world.entity.item.ItemEntity droppedItem = new net.minecraft.world.entity.item.ItemEntity(world, pos.getX() + 0.5D, pos.getY() + 0.5D, pos.getZ() + 0.5D, itemstack1.split(world.random.nextInt(21) + 10));
+                            if (flag1)
+                                droppedItem.setDeltaMovement(world.random.triangle(0.0D, 0.11485000171139836D), world.random.triangle(0.2D, 0.11485000171139836D), world.random.triangle(0.0D, 0.11485000171139836D));
+                            else
+                                droppedItem.setDeltaMovement(world.random.nextGaussian() * 0.05D, world.random.nextGaussian() * 0.05D + 0.2D, world.random.nextGaussian() * 0.05D);
+                            world.addFreshEntity(droppedItem);
+                            // Paper end
+                        }
+                        // Andromeda end
                         campfire.items.set(i, ItemStack.EMPTY);
                         world.sendBlockUpdated(pos, state, state, 3);
                         world.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(state));
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index b54295ec5d3130a430d657e9b43eff44053e42a8..a9fc869ff2cb7585cc4cb315811570c79641d9a8 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -135,4 +135,11 @@ public class LevelConfigurations extends ConfigurationPart {
 
     }
 
+    public PaperPullRequestsPort paperPullRequestsPort;
+    public class PaperPullRequestsPort extends ConfigurationPart {
+
+        public boolean fixCampfireRecipesNotAlwaysOutputtingFullResult = true;
+
+    }
+
 }
