From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 18:42:00 +0900
Subject: [PATCH] Port PurpurMC/Purpur#1234


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 83dde93586b0051fbfdff8051490b0dd46cd9913..b1051252c115852844245f6685f8a6b43e91b185 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -140,6 +140,23 @@ public class Explosion {
         if (this.radius < 0.1F) {
             return;
         }
+
+        // Andromeda start - Port PurpurMC/Purpur#1234
+        Location loc = new Location(this.level.getWorld(), this.x, this.y, this.z);
+        float yield = this.blockInteraction == BlockInteraction.DESTROY_WITH_DECAY ? 1.0F / this.radius : 1.0F;
+        if (this.source != null) {
+            if (!(new org.purpurmc.purpur.event.entity.PreEntityExplodeEvent(this.source.getBukkitEntity(), loc, yield).callEvent())) {
+                this.wasCanceled = true;
+                return;
+            }
+        } else {
+            if (!(new org.purpurmc.purpur.event.PreBlockExplodeEvent(loc.getBlock(), yield).callEvent())) {
+                this.wasCanceled = true;
+                return;
+            }
+        }
+        // Andromeda end
+
         // CraftBukkit end
         this.level.gameEvent(this.source, GameEvent.EXPLODE, new Vec3(this.x, this.y, this.z));
         Set<BlockPos> set = Sets.newHashSet();
