From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 19:41:03 +0900
Subject: [PATCH] Port PaperMC/Paper#8750


diff --git a/src/main/java/net/minecraft/world/inventory/MerchantResultSlot.java b/src/main/java/net/minecraft/world/inventory/MerchantResultSlot.java
index 7acf5ab9339e0134819aab5f270e99e927cc7a62..234e68c6ec5e459d7e2807b0ddf66849ab7bbbb9 100644
--- a/src/main/java/net/minecraft/world/inventory/MerchantResultSlot.java
+++ b/src/main/java/net/minecraft/world/inventory/MerchantResultSlot.java
@@ -64,6 +64,13 @@ public class MerchantResultSlot extends Slot {
                     return;
                 }
                 merchantOffer = org.bukkit.craftbukkit.inventory.CraftMerchantRecipe.fromBukkit(event.getTrade()).toMinecraft();
+                // Andromeda start - Port PaperMC/Paper#8750
+                if (player.getLevel().andromedaLevelConfiguration().paperPullRequestsPort.makePlayerPurchaseEventRespectResultItem) {
+                    stack.setItem(merchantOffer.result.getItem());
+                    stack.setTag(merchantOffer.result.getTag());
+                    stack.setCount(merchantOffer.result.getCount());
+                }
+                // Andromeda end
             }
         }
         this.checkTakeAchievements(stack);
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index a9fc869ff2cb7585cc4cb315811570c79641d9a8..ee3b28d5d6991212852f62b15616942a26be6618 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -140,6 +140,8 @@ public class LevelConfigurations extends ConfigurationPart {
 
         public boolean fixCampfireRecipesNotAlwaysOutputtingFullResult = true;
 
+        public boolean makePlayerPurchaseEventRespectResultItem = true;
+
     }
 
 }
