From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 19:41:03 +0900
Subject: [PATCH] Port PaperMC/Paper#8750


diff --git a/src/main/java/net/minecraft/world/inventory/MerchantResultSlot.java b/src/main/java/net/minecraft/world/inventory/MerchantResultSlot.java
index 7acf5ab9339e0134819aab5f270e99e927cc7a62..e4564cc37007134af2e87f3fca2d9b54b8177601 100644
--- a/src/main/java/net/minecraft/world/inventory/MerchantResultSlot.java
+++ b/src/main/java/net/minecraft/world/inventory/MerchantResultSlot.java
@@ -64,6 +64,11 @@ public class MerchantResultSlot extends Slot {
                     return;
                 }
                 merchantOffer = org.bukkit.craftbukkit.inventory.CraftMerchantRecipe.fromBukkit(event.getTrade()).toMinecraft();
+                // Andromeda start - Port PaperMC/Paper#8750
+                stack.setItem(merchantOffer.result.getItem());
+                stack.setTag(merchantOffer.result.getTag());
+                stack.setCount(merchantOffer.result.getCount());
+                // Andromeda end
             }
         }
         this.checkTakeAchievements(stack);
