From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 19:50:12 +0900
Subject: [PATCH] Port PaperMC/Paper#8744


diff --git a/src/main/java/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
index 0554afa72ae6d2f01886916031c8b90b0d45eb52..c1cd4b873a3556be7b63660634f7409e7149cd49 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
@@ -198,6 +198,7 @@ public class SculkShriekerBlockEntity extends BlockEntity implements VibrationLi
     }
 
     private boolean trySummonWarden(ServerLevel world) {
+        if (world.andromedaLevelConfiguration().paperPullRequestsPort.setWardenSpawnReasonToNatural.enabled) return this.warningLevel >= 4 && SpawnUtil.trySpawnMob(EntityType.WARDEN, MobSpawnType.TRIGGERED, world, this.getBlockPos(), 20, 5, 6, SpawnUtil.Strategy.ON_TOP_OF_COLLIDER, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL, null).isPresent(); // Andromeda - Port PaperMC/Paper#8744
         return this.warningLevel < 4 ? false : SpawnUtil.trySpawnMob(EntityType.WARDEN, MobSpawnType.TRIGGERED, world, this.getBlockPos(), 20, 5, 6, SpawnUtil.Strategy.ON_TOP_OF_COLLIDER).isPresent();
     }
 
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 7e07776405c5c2a5f8be6a993d63f795e2b41909..9207f0db9ef54b3c075df60dc32097bbfa34649b 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -116,6 +116,13 @@ public class LevelConfigurations extends ConfigurationPart {
 
         }
 
+        public SetWardenSpawnReasonToNatural setWardenSpawnReasonToNatural;
+        public class SetWardenSpawnReasonToNatural extends ConfigurationPart {
+
+            public boolean enabled = true;
+
+        }
+
     }
 
 }
