From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 22:07:46 +0900
Subject: [PATCH] Port PaperMC/Paper#8733


diff --git a/src/main/java/io/papermc/paper/voxel/AABBVoxelShape.java b/src/main/java/io/papermc/paper/voxel/AABBVoxelShape.java
index d67a40e7be030142443680c89e1763fc9ecdfe0a..ecfaed222861208e58a2f42a2567d094d5bfb591 100644
--- a/src/main/java/io/papermc/paper/voxel/AABBVoxelShape.java
+++ b/src/main/java/io/papermc/paper/voxel/AABBVoxelShape.java
@@ -178,7 +178,7 @@ public final class AABBVoxelShape extends VoxelShape {
 
     @Override
     public double collide(Direction.Axis enumdirection_enumaxis, AABB axisalignedbb, double d0) {
-        if (CollisionUtil.isEmpty(this.aabb) || CollisionUtil.isEmpty(axisalignedbb)) {
+        if (CollisionUtil.isEmpty(this.aabb)) { // Andromeda - Port PaperMC/Paper#8733
             return d0;
         }
         switch (enumdirection_enumaxis.ordinal()) {
