From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 22:08:21 +0900
Subject: [PATCH] Port PaperMC/Paper#8713


diff --git a/src/main/java/net/minecraft/world/item/HoneycombItem.java b/src/main/java/net/minecraft/world/item/HoneycombItem.java
index 1f8b7b50c6aa24778d87821ae2ff4d019d176082..27abaf5f323d554dab5c03a06c7c6257d7d9413a 100644
--- a/src/main/java/net/minecraft/world/item/HoneycombItem.java
+++ b/src/main/java/net/minecraft/world/item/HoneycombItem.java
@@ -39,6 +39,7 @@ public class HoneycombItem extends Item {
             ItemStack itemStack = context.getItemInHand();
             // Paper start - EntityChangeBlockEvent
             if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(player, blockPos, state).isCancelled()) {
+                if (!player.isCreative() && level.andromedaLevelConfiguration().paperPullRequestsPort.fixDesyncOfHoneycombWhenTheEventIsCancelled.enabled) player.containerMenu.sendAllDataToRemote(); // Andromeda - Port PaperMC/Paper#8713
                 return InteractionResult.PASS;
             }
             // Paper end
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 2594aabb99c8deb15f4f38544fd1b7baca2dfbae..6adead94485ad859f8566be56f0e5dc4fbe70ef7 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -159,6 +159,13 @@ public class LevelConfigurations extends ConfigurationPart {
 
         }
 
+        public FixDesyncOfHoneycombWhenTheEventIsCancelled fixDesyncOfHoneycombWhenTheEventIsCancelled;
+        public class FixDesyncOfHoneycombWhenTheEventIsCancelled extends ConfigurationPart {
+
+            public boolean enabled = true;
+
+        }
+
     }
 
 }
