From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 22:46:29 +0900
Subject: [PATCH] Port PaperMC/Paper#8657


diff --git a/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java b/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
index 88ce7d0c049cefdfa99a21ddecc5d16e9435a4f0..4cb6f7f54731feb43e2990b27b5c5060a351a8c3 100644
--- a/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -756,6 +756,13 @@ public interface DispenseItemBehavior {
             protected ItemStack execute(BlockSource pointer, ItemStack stack) {
                 ServerLevel worldserver = pointer.getLevel();
 
+                // Andromeda start - Port PaperMC/Paper#8657
+                Direction enumdirection = (Direction) pointer.getBlockState().getValue(DispenserBlock.FACING);
+                BlockPos blockposition = pointer.getPos().relative(enumdirection);
+                BlockState iblockdata = worldserver.getBlockState(blockposition);
+                if (BaseFireBlock.canBePlacedAt(worldserver, blockposition, enumdirection) || CampfireBlock.canLight(iblockdata) || CandleBlock.canLight(iblockdata) || CandleCakeBlock.canLight(iblockdata) || iblockdata.getBlock() instanceof TntBlock || !worldserver.andromedaLevelConfiguration().paperPullRequestsPort.fixBlockDispenseEventBeingCalledWhenNothingActuallyHappens) {
+                // Andromeda end
+
                 // CraftBukkit start
                 org.bukkit.block.Block bukkitBlock = worldserver.getWorld().getBlockAt(pointer.getPos().getX(), pointer.getPos().getY(), pointer.getPos().getZ());
                 CraftItemStack craftItem = CraftItemStack.asCraftMirror(stack); // Paper - ignore stack size on damageable items
@@ -779,11 +786,9 @@ public interface DispenseItemBehavior {
                     }
                 }
                 // CraftBukkit end
+                } // Andromeda - Port PaperMC/Paper#8657
 
                 this.setSuccess(true);
-                Direction enumdirection = (Direction) pointer.getBlockState().getValue(DispenserBlock.FACING);
-                BlockPos blockposition = pointer.getPos().relative(enumdirection);
-                BlockState iblockdata = worldserver.getBlockState(blockposition);
 
                 if (BaseFireBlock.canBePlacedAt(worldserver, blockposition, enumdirection)) {
                     // CraftBukkit start - Ignition by dispensing flint and steel
@@ -817,6 +822,10 @@ public interface DispenseItemBehavior {
                 this.setSuccess(true);
                 ServerLevel worldserver = pointer.getLevel();
                 BlockPos blockposition = pointer.getPos().relative((Direction) pointer.getBlockState().getValue(DispenserBlock.FACING));
+                // Andromeda start - Port PaperMC/Paper#8657
+                final BlockState blockState = worldserver.getBlockState(blockposition);
+                if ((blockState.getBlock() instanceof net.minecraft.world.level.block.BonemealableBlock bonemealableBlock && bonemealableBlock.isValidBonemealTarget(worldserver, blockposition, blockState, worldserver.isClientSide)) || (blockState.is(Blocks.WATER) && blockState.getFluidState().getAmount() == 8) || !worldserver.andromedaLevelConfiguration().paperPullRequestsPort.fixBlockDispenseEventBeingCalledWhenNothingActuallyHappens) {
+                // Andromeda end
                 // CraftBukkit start
                 org.bukkit.block.Block block = worldserver.getWorld().getBlockAt(pointer.getPos().getX(), pointer.getPos().getY(), pointer.getPos().getZ());
                 CraftItemStack craftItem = CraftItemStack.asCraftMirror(stack.copyWithCount(1)); // Paper - single item in event
@@ -839,6 +848,7 @@ public interface DispenseItemBehavior {
                         return stack;
                     }
                 }
+                } // Andromeda - Port PaperMC/Paper#8657
 
                 worldserver.captureTreeGeneration = true;
                 // CraftBukkit end
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 2fa13e6815fcff03a45f1885fbf87e2504376537..a95b4405a57da140409dbd42c78cbd88136d391a 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -146,6 +146,8 @@ public class LevelConfigurations extends ConfigurationPart {
 
         public boolean fixSeveralIssuesWithEntityBreedEvent = true;
 
+        public boolean fixBlockDispenseEventBeingCalledWhenNothingActuallyHappens = true;
+
     }
 
 }
