From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 22:35:15 +0900
Subject: [PATCH] Port PaperMC/Paper#8677


diff --git a/src/main/java/net/minecraft/world/entity/animal/Animal.java b/src/main/java/net/minecraft/world/entity/animal/Animal.java
index 2ac88f06ebb79e515cd9934ac1e3e2c8003d9e3c..5c263afce61e1a41a0207ff1c66a9f447b391e62 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Animal.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Animal.java
@@ -152,6 +152,7 @@ public abstract class Animal extends AgeableMob {
             int i = this.getAge();
 
             if (!this.level.isClientSide && i == 0 && this.canFallInLove() && (this.level.purpurConfig.animalBreedingCooldownSeconds <= 0 || !this.level.hasBreedingCooldown(player.getUUID(), this.getClass()))) { // Purpur
+                if (this.level.andromedaLevelConfiguration().paperPullRequestsPort.fixSeveralIsuesWithEntityBreedEvent) this.breedItem = itemstack.copy(); // Andromeda - Port PaperMC/Paper#8677
                 this.usePlayerItem(player, hand, itemstack);
                 this.setInLove(player);
                 return InteractionResult.SUCCESS;
@@ -186,6 +187,7 @@ public abstract class Animal extends AgeableMob {
         // CraftBukkit start
         EntityEnterLoveModeEvent entityEnterLoveModeEvent = CraftEventFactory.callEntityEnterLoveModeEvent(player, this, 600);
         if (entityEnterLoveModeEvent.isCancelled()) {
+            if (this.level.andromedaLevelConfiguration().paperPullRequestsPort.fixSeveralIsuesWithEntityBreedEvent) this.breedItem = null; // Andromeda - Port PaperMC/Paper#8677
             return;
         }
         this.inLove = entityEnterLoveModeEvent.getTicksInLove();
@@ -193,7 +195,7 @@ public abstract class Animal extends AgeableMob {
         if (player != null) {
             this.loveCause = player.getUUID();
         }
-        this.breedItem = player.getInventory().getSelected(); // CraftBukkit
+        if (!this.level.andromedaLevelConfiguration().paperPullRequestsPort.fixSeveralIsuesWithEntityBreedEvent) this.breedItem = player.getInventory().getSelected(); // Andromeda - Port PaperMC/Paper#8677
 
         this.level.broadcastEntityEvent(this, (byte) 18);
     }
diff --git a/src/main/java/net/minecraft/world/entity/animal/Panda.java b/src/main/java/net/minecraft/world/entity/animal/Panda.java
index 841838562ffed67127b03e27f61d692d9933fbe3..519136a8a63332a280d7a53c0c6a76ea4265ebd1 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Panda.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Panda.java
@@ -695,6 +695,7 @@ public class Panda extends Animal {
                 this.usePlayerItem(player, hand, itemstack);
                 this.ageUp((int) ((float) (-this.getAge() / 20) * 0.1F), true);
             } else if (!this.level.isClientSide && this.getAge() == 0 && this.canFallInLove()) {
+                if (this.level.andromedaLevelConfiguration().paperPullRequestsPort.fixSeveralIsuesWithEntityBreedEvent) this.breedItem = itemstack.copy(); // Andromeda - Port PaperMC/Paper#8677
                 this.usePlayerItem(player, hand, itemstack);
                 this.setInLove(player);
             } else {
diff --git a/src/main/java/net/minecraft/world/entity/animal/camel/Camel.java b/src/main/java/net/minecraft/world/entity/animal/camel/Camel.java
index 7f3d93a3c41c1e28afba1051fe7403a1fa045a2d..64090385fd61bac8c4b72ff11f62ebd71f641661 100644
--- a/src/main/java/net/minecraft/world/entity/animal/camel/Camel.java
+++ b/src/main/java/net/minecraft/world/entity/animal/camel/Camel.java
@@ -419,6 +419,7 @@ public class Camel extends AbstractHorse implements PlayerRideableJumping, Rider
 
             boolean bl2 = this.isTamed() && this.getAge() == 0 && this.canFallInLove();
             if (bl2) {
+                if (this.level.andromedaLevelConfiguration().paperPullRequestsPort.fixSeveralIsuesWithEntityBreedEvent) this.breedItem = item.copy(); // Andromeda - Port PaperMC/Paper#8677
                 this.setInLove(player);
             }
 
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java b/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
index 8f27e6b495b82361d331c514c78088d358e4f5b4..97f042831147a696ac6e5baa278c0faf555460f7 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
@@ -519,6 +519,7 @@ public abstract class AbstractHorse extends Animal implements ContainerListener,
             b0 = 5;
             if (!this.level.isClientSide && this.isTamed() && this.getAge() == 0 && !this.isInLove()) {
                 flag = true;
+                if (this.level.andromedaLevelConfiguration().paperPullRequestsPort.fixSeveralIsuesWithEntityBreedEvent) this.breedItem = item.copy(); // Andromeda - Port PaperMC/Paer#8677
                 this.setInLove(player);
             }
         } else if (item.is(Items.GOLDEN_APPLE) || item.is(Items.ENCHANTED_GOLDEN_APPLE)) {
@@ -527,6 +528,7 @@ public abstract class AbstractHorse extends Animal implements ContainerListener,
             b0 = 10;
             if (!this.level.isClientSide && this.isTamed() && this.getAge() == 0 && !this.isInLove()) {
                 flag = true;
+                if (this.level.andromedaLevelConfiguration().paperPullRequestsPort.fixSeveralIsuesWithEntityBreedEvent) this.breedItem = item.copy(); // Andromeda - Port PaperMC/Paer#8677
                 this.setInLove(player);
             }
         }
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java b/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
index beea0545a38b0f044409c2cdb5bbefaf8d783d45..df924148520cab16aed84657459c9b7bd440c0a5 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
@@ -289,6 +289,7 @@ public class Llama extends AbstractChestedHorse implements VariantHolder<Llama.V
             f = 10.0F;
             if (this.isTamed() && this.getAge() == 0 && this.canFallInLove()) {
                 flag = true;
+                if (this.level.andromedaLevelConfiguration().paperPullRequestsPort.fixSeveralIsuesWithEntityBreedEvent) this.breedItem = item.copy(); // Andromeda - Port PaperMC/Paer#8677
                 this.setInLove(player);
             }
         }
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 125a41dc84df3f61aed6ffd03ba2c6d3ade4e906..b54347d0818bf8beed4812331cf7cc8bf24675f7 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -146,6 +146,8 @@ public class LevelConfigurations extends ConfigurationPart {
 
         public boolean fixDesyncOfHoneycombWhenTheEventIsCancelled = true;
 
+        public boolean fixSeveralIsuesWithEntityBreedEvent = true;
+
     }
 
 }
