From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 22:46:29 +0900
Subject: [PATCH] Port PaperMC/Paper#8657


diff --git a/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java b/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
index 2df7bb73a004c9b3f1e5eaff2557e3cfafb9b3fb..98f3711886361077015ba6e08f4566c5b01846a5 100644
--- a/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -756,6 +756,13 @@ public interface DispenseItemBehavior {
             protected ItemStack execute(BlockSource pointer, ItemStack stack) {
                 ServerLevel worldserver = pointer.getLevel();
 
+                // Andromeda start - Port PaperMC/Paper#8657
+                Direction enumdirection = (Direction) pointer.getBlockState().getValue(DispenserBlock.FACING);
+                BlockPos blockposition = pointer.getPos().relative(enumdirection);
+                BlockState iblockdata = worldserver.getBlockState(blockposition);
+                if (BaseFireBlock.canBePlacedAt(worldserver, blockposition, enumdirection) || CampfireBlock.canLight(iblockdata) || CandleBlock.canLight(iblockdata) || CandleCakeBlock.canLight(iblockdata) || iblockdata.getBlock() instanceof TntBlock || !worldserver.andromedaLevelConfiguration().paperPullRequestsPort.fixBlockDispenseEventBeingCalledWhenNothingActuallyHappens.enabled) {
+                // Andromeda end
+
                 // CraftBukkit start
                 org.bukkit.block.Block bukkitBlock = worldserver.getWorld().getBlockAt(pointer.getPos().getX(), pointer.getPos().getY(), pointer.getPos().getZ());
                 CraftItemStack craftItem = CraftItemStack.asCraftMirror(stack); // Paper - ignore stack size on damageable items
@@ -779,11 +786,9 @@ public interface DispenseItemBehavior {
                     }
                 }
                 // CraftBukkit end
+                } // Andromeda - Port PaperMC/Paper#8657
 
                 this.setSuccess(true);
-                Direction enumdirection = (Direction) pointer.getBlockState().getValue(DispenserBlock.FACING);
-                BlockPos blockposition = pointer.getPos().relative(enumdirection);
-                BlockState iblockdata = worldserver.getBlockState(blockposition);
 
                 if (BaseFireBlock.canBePlacedAt(worldserver, blockposition, enumdirection)) {
                     // CraftBukkit start - Ignition by dispensing flint and steel
@@ -817,6 +822,10 @@ public interface DispenseItemBehavior {
                 this.setSuccess(true);
                 ServerLevel worldserver = pointer.getLevel();
                 BlockPos blockposition = pointer.getPos().relative((Direction) pointer.getBlockState().getValue(DispenserBlock.FACING));
+                // Andromeda start - Port PaperMC/Paper#8657
+                final BlockState blockState = worldserver.getBlockState(blockposition);
+                if ((blockState.getBlock() instanceof net.minecraft.world.level.block.BonemealableBlock bonemealableBlock && bonemealableBlock.isValidBonemealTarget(worldserver, blockposition, blockState, worldserver.isClientSide)) || (blockState.is(Blocks.WATER) && blockState.getFluidState().getAmount() == 8) || !worldserver.andromedaLevelConfiguration().paperPullRequestsPort.fixBlockDispenseEventBeingCalledWhenNothingActuallyHappens.enabled) {
+                // Andromeda end
                 // CraftBukkit start
                 org.bukkit.block.Block block = worldserver.getWorld().getBlockAt(pointer.getPos().getX(), pointer.getPos().getY(), pointer.getPos().getZ());
                 CraftItemStack craftItem = CraftItemStack.asCraftMirror(stack.copyWithCount(1)); // Paper - single item in event
@@ -839,6 +848,7 @@ public interface DispenseItemBehavior {
                         return stack;
                     }
                 }
+                } // Andromeda - Port PaperMC/Paper#8657
 
                 worldserver.captureTreeGeneration = true;
                 // CraftBukkit end
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 9eb40feb1870c1601c7be8ab573f902e7f26d55f..042e25b01e2fa78aa8c79865a1d057f354a6170f 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -173,6 +173,13 @@ public class LevelConfigurations extends ConfigurationPart {
 
         }
 
+        public FixBlockDispenseEventBeingCalledWhenNothingActuallyHappens fixBlockDispenseEventBeingCalledWhenNothingActuallyHappens;
+        public class FixBlockDispenseEventBeingCalledWhenNothingActuallyHappens extends ConfigurationPart {
+
+            public boolean enabled = true;
+
+        }
+
     }
 
 }
