From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 22:58:54 +0900
Subject: [PATCH] Port PaperMC/Paper#8602


diff --git a/src/main/java/net/minecraft/commands/CommandSourceStack.java b/src/main/java/net/minecraft/commands/CommandSourceStack.java
index 613b3fb0a90ef19161f0df09a014c3811e5b8c9e..9deffdc96469d213120fb08723644cf067639223 100644
--- a/src/main/java/net/minecraft/commands/CommandSourceStack.java
+++ b/src/main/java/net/minecraft/commands/CommandSourceStack.java
@@ -207,8 +207,18 @@ public class CommandSourceStack implements SharedSuggestionProvider, com.destroy
 
     // CraftBukkit start
     public boolean hasPermission(int i, String bukkitPermission) {
-        // World is null when loading functions
-        return ((this.getLevel() == null || !this.getLevel().getCraftServer().ignoreVanillaPermissions) && this.permissionLevel >= i) || this.getBukkitSender().hasPermission(bukkitPermission);
+        // Andromeda start - Port PaperMC/Paper#8602
+        boolean flag = true;
+        try {
+            flag = team.earendel.andromeda.configurations.GlobalConfiguration.get().paperPullRequestsPort.fixCommandFunctionPermChecking;
+        } catch (Exception ignored) {;}
+        if (!flag)
+            return ((this.getLevel() == null || !this.getLevel().getCraftServer().ignoreVanillaPermissions) && this.permissionLevel >= i) || this.getBukkitSender().hasPermission(bukkitPermission);
+        if (this.source == CommandSource.NULL)
+            return this.permissionLevel >= i;
+        else
+            return (!this.getLevel().getCraftServer().ignoreVanillaPermissions && this.permissionLevel >= i) || this.getBukkitSender().hasPermission(bukkitPermission);
+        // Andromeda end
     }
     // CraftBukkit end
 
diff --git a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
index 09b8e1b0732dd74cc38152b8cc402cc1ad356242..1231d8f4ac9c73679e46f19c5a6bcbb479fbfbb2 100644
--- a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
+++ b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
@@ -160,6 +160,8 @@ public class GlobalConfiguration extends ConfigurationPart {
 
         public boolean properlyTrackTheChangedItemFromDispenseEvents = true;
 
+        public boolean fixCommandFunctionPermChecking = true;
+
     }
 
 }
