From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 22:59:41 +0900
Subject: [PATCH] Port PaperMC/Paper#8600


diff --git a/src/main/java/net/minecraft/world/item/RecordItem.java b/src/main/java/net/minecraft/world/item/RecordItem.java
index 4c6f60cc1269ed0315b7e5a54235b9c95d1f039c..a6a991ce640620f80ff0649a015c6fffb4b2e7d7 100644
--- a/src/main/java/net/minecraft/world/item/RecordItem.java
+++ b/src/main/java/net/minecraft/world/item/RecordItem.java
@@ -43,7 +43,11 @@ public class RecordItem extends Item {
             ItemStack itemstack = context.getItemInHand();
 
             if (!world.isClientSide) {
-                if (true) return InteractionResult.SUCCESS; // CraftBukkit - handled in ItemStack
+                // Andromeda start - Port PaperMC/Paper#8600
+                if (world.andromedaLevelConfiguration().paperPullRequestsPort.removeDuplicateAnimatePacketForRecords) return InteractionResult.sidedSuccess(world.isClientSide); // CraftBukkit - handled in ItemStack
+                else return InteractionResult.SUCCESS;
+                /*
+                // Andromeda end
                 ((JukeboxBlock) Blocks.JUKEBOX).setRecord(context.getPlayer(), world, blockposition, iblockdata, itemstack);
                 world.levelEvent((Player) null, 1010, blockposition, Item.getId(this));
                 itemstack.shrink(1);
@@ -52,6 +56,7 @@ public class RecordItem extends Item {
                 if (entityhuman != null) {
                     entityhuman.awardStat(Stats.PLAY_RECORD);
                 }
+                 */ // Andromeda
             }
 
             return InteractionResult.sidedSuccess(world.isClientSide);
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index ff2188562877f036c028c2f7c5326b2e408ca581..49dcc3468d8e3308e74d2f6df9364045476a91f5 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -152,6 +152,8 @@ public class LevelConfigurations extends ConfigurationPart {
 
         public boolean includePreviousSpawnAngleInSpawnChangeEvent = true;
 
+        public boolean removeDuplicateAnimatePacketForRecords = true;
+
     }
 
 }
