From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 7 Jan 2023 23:04:48 +0900
Subject: [PATCH] Port PaperMC/Paper#8534


diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
index 50b691fd4086a9b049c83936abfee9ae0414837d..1c01d1c4d4e67233eccfe7fb70b1949a2b3dd888 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
@@ -8,6 +8,7 @@ import net.minecraft.network.protocol.Packet;
 // Spigot start
 public record ClientboundSystemChatPacket(@javax.annotation.Nullable net.kyori.adventure.text.Component adventure$content, @javax.annotation.Nullable String content, boolean overlay) implements Packet<ClientGamePacketListener> { // Paper - Adventure
 
+    @io.papermc.paper.annotation.DoNotUse // Andromeda - Port PaperMC/Paper#8534
     public ClientboundSystemChatPacket(Component content, boolean overlay) {
         this(null, Component.Serializer.toJson(content), overlay); // Paper - Adventure
     }
@@ -35,7 +36,7 @@ public record ClientboundSystemChatPacket(@javax.annotation.Nullable net.kyori.a
     // Paper end
 
     public ClientboundSystemChatPacket(FriendlyByteBuf buf) {
-        this(buf.readComponent(), buf.readBoolean());
+        this(null, io.papermc.paper.adventure.PaperAdventure.asJsonString(buf.readComponent(), buf.adventure$locale), buf.readBoolean()); // Andromeda - Port PaperMC/Paper#8534
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 5968c3b2c9e4871691ba9a090ea14122023fe48c..af7e769755d2897226730025aee55287d9537a46 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1974,13 +1974,13 @@ public class ServerPlayer extends Player {
 
     public void sendSystemMessage(Component message, boolean overlay) {
         if (this.acceptsSystemMessages(overlay)) {
-            this.connection.send(new ClientboundSystemChatPacket(message, overlay), PacketSendListener.exceptionallySend(() -> {
+            this.connection.send(new ClientboundSystemChatPacket(PaperAdventure.asAdventure(message), overlay), PacketSendListener.exceptionallySend(() -> { // Andromeda - Port PaperMC/Paper#8534
                 if (this.acceptsSystemMessages(false)) {
                     boolean flag1 = true;
                     String s = message.getString(256);
                     MutableComponent ichatmutablecomponent = Component.literal(s).withStyle(ChatFormatting.YELLOW);
 
-                    return new ClientboundSystemChatPacket(Component.translatable("multiplayer.message_not_delivered", ichatmutablecomponent).withStyle(ChatFormatting.RED), false);
+                    return new ClientboundSystemChatPacket(PaperAdventure.asAdventure(Component.translatable("multiplayer.message_not_delivered", ichatmutablecomponent).withStyle(ChatFormatting.RED)), false); // Andromeda - Port PaperMC/Paper#8534
                 } else {
                     return null;
                 }
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 6bde9706c48030fa6b2b92f9376eaa8cb945682c..123f4bb857483b3f6e31acd49739a8f40c4d8fd3 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2237,9 +2237,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
                 LOGGER.info("Sending message: {}", chat.unsignedContent() != null ? chat.unsignedContent() : chat.body().content());
 
             if (team.earendel.andromeda.configurations.AndromedaConfiguration.NCRConvertToGameMessage) {
-                packet = new ClientboundSystemChatPacket(
-                        chat.chatType().resolve(this.player.level.registryAccess())
-                                .get().decorate(chat.unsignedContent() != null ? chat.unsignedContent() : Component.literal(chat.body().content())), false);
+                packet = new ClientboundSystemChatPacket(PaperAdventure.asAdventure(chat.chatType().resolve(this.player.level.registryAccess()).get().decorate(chat.unsignedContent() != null ? chat.unsignedContent() : Component.literal(chat.body().content()))), false);
                 this.send(packet, null);
                 return;
             }
@@ -2443,7 +2441,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
             }); // Paper - push to main
             return Optional.empty();
         } else if (this.player.isRemoved() || this.player.getChatVisibility() == ChatVisiblity.HIDDEN) { // CraftBukkit - dead men tell no tales
-            this.send(new ClientboundSystemChatPacket(Component.translatable("chat.disabled.options").withStyle(ChatFormatting.RED), false));
+            this.send(new ClientboundSystemChatPacket(PaperAdventure.asAdventure(Component.translatable("chat.disabled.options").withStyle(ChatFormatting.RED)), false)); // Andromeda - Port PaperMC/Paper#8534
             return Optional.empty();
         } else {
             Optional<LastSeenMessages> optional = this.unpackAndApplyLastSeen(acknowledgment);
@@ -2660,7 +2658,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
                 }
             });
         } else if (this.player.getChatVisibility() == ChatVisiblity.SYSTEM) { // Re-add "Command Only" flag check
-            this.send(new ClientboundSystemChatPacket(Component.translatable("chat.cannotSend").withStyle(ChatFormatting.RED), false));
+            this.send(new ClientboundSystemChatPacket(PaperAdventure.asAdventure(Component.translatable("chat.cannotSend").withStyle(ChatFormatting.RED)), false)); // Andromeda - Port PaperMC/Paper#8534
         } else {
             this.chat(s, message, true);
         }
