From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sun, 8 Jan 2023 11:41:16 +0900
Subject: [PATCH] Port PaperMC/Paper#8514


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 07b0beebdcc0fff0b8403dc1c309d41c4a4260f6..3543ddf7c9d2648136025407d1d4d545cda71cc0 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1201,6 +1201,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         targetLocation.setPitch(pitch);
 
         org.bukkit.util.Vector direction = targetLocation.getDirection();
+        if (this.getHandle().getLevel().andromedaLevelConfiguration().paperPullRequestsPort.moveBlockFartherAwayForPlayerSetRotation) direction.multiply(9999999); // Andromeda - Port PaperMC/Paper#8514
         targetLocation.add(direction);
         this.lookAt(targetLocation, io.papermc.paper.entity.LookAnchor.EYES);
         // Paper end
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index a654834d93542905c6d102dfa76e3d4d2553f197..33a3a7d00bc4795d1cdd4f3268e71b62a8bf076e 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -156,6 +156,8 @@ public class LevelConfigurations extends ConfigurationPart {
 
         public boolean removeDuplicateAnimatePacketForRecords = true;
 
+        public boolean moveBlockFartherAwayForPlayerSetRotation = true;
+
     }
 
 }
