From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sun, 8 Jan 2023 11:46:36 +0900
Subject: [PATCH] Port PaperMC/Paper#8505


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 6bf14183a3fcd2b3d166752ce33240d2ff1ffa7c..57b7ff19d20f7798b75706b4b65119f40bd8af1e 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -154,6 +154,16 @@ public class GlobalConfiguration extends ConfigurationPart {
         public boolean allowHeadlessPistons = false;
         @Comment("This setting controls if grindstones should be able to output overstacked items (such as cursed books).")
         public boolean allowGrindstoneOverstacking = false;
+        // Andromeda start - Port PaperMC/Paper#8505
+        @Comment("This setting controls what compression format is used for region files.")
+        public CompressionFormat compressionFormat = CompressionFormat.ZLIB;
+
+        public enum CompressionFormat {
+            GZIP,
+            ZLIB,
+            NONE
+        }
+        // Andromeda end
     }
 
     public Commands commands;
diff --git a/src/main/java/net/minecraft/world/level/chunk/storage/RegionFile.java b/src/main/java/net/minecraft/world/level/chunk/storage/RegionFile.java
index ef797d47c6bbc1c1e3ff6c15c04e1df41f727436..c294ac004cc9e085e56c55c27717deadacd93624 100644
--- a/src/main/java/net/minecraft/world/level/chunk/storage/RegionFile.java
+++ b/src/main/java/net/minecraft/world/level/chunk/storage/RegionFile.java
@@ -426,11 +426,11 @@ public class RegionFile implements AutoCloseable {
     // Paper end
 
     public RegionFile(Path file, Path directory, boolean dsync) throws IOException {
-        this(file, directory, RegionFileVersion.VERSION_DEFLATE, dsync);
+        this(file, directory, RegionFileVersion.COMPRESSION_FORMAT_SUPPLIER.get(), dsync); // Andromeda - Port PaperMC/Paper#8505
     }
     // Paper start - add can recalc flag
     public RegionFile(Path file, Path directory, boolean dsync, boolean canRecalcHeader) throws IOException {
-        this(file, directory, RegionFileVersion.VERSION_DEFLATE, dsync, canRecalcHeader);
+        this(file, directory, RegionFileVersion.COMPRESSION_FORMAT_SUPPLIER.get(), dsync, canRecalcHeader); // Andromeda - Port PaperMC/Paper#8505
     }
     // Paper end - add can recalc flag
 
diff --git a/src/main/java/net/minecraft/world/level/chunk/storage/RegionFileVersion.java b/src/main/java/net/minecraft/world/level/chunk/storage/RegionFileVersion.java
index 4411e427d3b6b592f8a18e61b6c59309cf699d3f..016c3b3f4b5edcaf102f86550652c4e12f4ae167 100644
--- a/src/main/java/net/minecraft/world/level/chunk/storage/RegionFileVersion.java
+++ b/src/main/java/net/minecraft/world/level/chunk/storage/RegionFileVersion.java
@@ -30,6 +30,14 @@ public class RegionFileVersion {
     }, (stream) -> {
         return stream;
     }));
+    // Andromeda start - Port PaperMC/Paper#8505
+    public static final java.util.function.Supplier<RegionFileVersion> COMPRESSION_FORMAT_SUPPLIER = () ->
+            switch (io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.compressionFormat) {
+                case GZIP -> VERSION_GZIP;
+                case ZLIB -> VERSION_DEFLATE;
+                case NONE -> VERSION_NONE;
+            };
+    // Andromeda end
     private final int id;
     private final RegionFileVersion.StreamWrapper<InputStream> inputWrapper;
     private final RegionFileVersion.StreamWrapper<OutputStream> outputWrapper;
