From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sun, 8 Jan 2023 11:49:49 +0900
Subject: [PATCH] Port PaperMC/Paper#8473


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 6112cc120e84271624bdf3d35ed94e8f663e74b2..16caddc8c056fddb32fe9a2f61e648cf77eaa9fe 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -988,7 +988,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
         if (this.getHandle().connection == null) return;
 
-        int stage = (int) (9 * progress); // There are 0 - 9 damage states
+        int stage = this.getHandle().level.andromedaLevelConfiguration().paperPullRequestsPort.fixSendBlockDamage0Progress.enabled ? (int) (progress * 10) - 1 : (int) progress * 9; // There are 0 - 9 damage states // Andromeda - Port PaperMC/Paper#8473
         ClientboundBlockDestructionPacket packet = new ClientboundBlockDestructionPacket(destroyerIdentity, new BlockPos(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()), stage);  // Paper - customBlockDamage identity
         this.getHandle().connection.send(packet);
     }
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 1d3077322ae0434439710dd7c8670da2208fa242..b3af23f43ac912b94e9de3bd1e735cc4b0059773 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -208,6 +208,13 @@ public class LevelConfigurations extends ConfigurationPart {
 
         }
 
+        public FixSendBlockDamage0Progress fixSendBlockDamage0Progress;
+        public class FixSendBlockDamage0Progress extends ConfigurationPart {
+
+            public boolean enabled = true;
+
+        }
+
     }
 
 }
