From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Tue, 10 Jan 2023 23:57:44 +0900
Subject: [PATCH] Add configuration to disable not secure chat marker


diff --git a/src/main/java/io/papermc/paper/adventure/ChatProcessor.java b/src/main/java/io/papermc/paper/adventure/ChatProcessor.java
index 309fe1162db195c7c3c94d785d6aa2700e42b08a..a7aec3c9c810b34cd20d883deaa5ffbf82f82c9b 100644
--- a/src/main/java/io/papermc/paper/adventure/ChatProcessor.java
+++ b/src/main/java/io/papermc/paper/adventure/ChatProcessor.java
@@ -352,7 +352,7 @@ public final class ChatProcessor {
 
         private void sendToServer(final ChatType.Bound chatType, final @Nullable Function<Audience, net.minecraft.network.chat.Component> msgFunction) {
             final PlayerChatMessage toConsoleMessage = msgFunction == null ? ChatProcessor.this.message : ChatProcessor.this.message.withUnsignedContent(msgFunction.apply(ChatProcessor.this.server.console));
-            ChatProcessor.this.server.logChatMessage(toConsoleMessage.decoratedContent(), chatType, ChatProcessor.this.server.getPlayerList().verifyChatTrusted(toConsoleMessage) ? null : "Not Secure");
+            ChatProcessor.this.server.logChatMessage(toConsoleMessage.decoratedContent(), chatType, ChatProcessor.this.server.getPlayerList().verifyChatTrusted(toConsoleMessage) || !team.earendel.andromeda.configurations.GlobalConfiguration.get().consoleLogs.notSecureChatMarker ? null : "Not Secure"); // Andromeda
         }
     }
 
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index c457f114c1ec2e402df65a17a838e165f6b1a31d..c9f69c0bf3ab4ba38f9a226e5cf18956012f0ea7 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1419,7 +1419,7 @@ public abstract class PlayerList {
     }
     public void broadcastChatMessage(PlayerChatMessage message, Predicate<ServerPlayer> shouldSendFiltered, @Nullable ServerPlayer sender, ChatType.Bound params, @Nullable Function<net.kyori.adventure.audience.Audience, Component> unsignedFunction) {
         // Paper end
-        boolean flag = this.verifyChatTrusted(message);
+        boolean flag = this.verifyChatTrusted(message) || !team.earendel.andromeda.configurations.GlobalConfiguration.get().consoleLogs.notSecureChatMarker; // Andromeda
 
         this.server.logChatMessage((unsignedFunction == null ? message.decoratedContent() : unsignedFunction.apply(this.server.console)), params, flag ? null : "Not Secure"); // Paper
         OutgoingChatMessage outgoingchatmessage = OutgoingChatMessage.create(message);
diff --git a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
index 89aa59ed398a4e5fac1218580dd7fb35a5c0c31e..9830e2ba3ad8e7cad3b489438066c14a5c617541 100644
--- a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
+++ b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
@@ -33,6 +33,7 @@ public class GlobalConfiguration extends ConfigurationPart {
 
         public boolean offlineWarnings = true;
         public boolean rootUserWarnings = true;
+        public boolean notSecureChatMarker = true;
 
     }
 
