From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sun, 8 Jan 2023 12:04:07 +0900
Subject: [PATCH] Port PaperMC/Paper#8418


diff --git a/src/main/java/net/minecraft/network/FriendlyByteBuf.java b/src/main/java/net/minecraft/network/FriendlyByteBuf.java
index 1b271896213f40b007a87ebbfe44e5c774db6994..92b30af59cc2e26b0a86721c3cdd88ba3d7b1b97 100644
--- a/src/main/java/net/minecraft/network/FriendlyByteBuf.java
+++ b/src/main/java/net/minecraft/network/FriendlyByteBuf.java
@@ -96,6 +96,20 @@ public class FriendlyByteBuf extends ByteBuf {
     }
 
     public static int getVarIntSize(int value) {
+        // Andromeda start - Port PaperMC/Paper#8414
+        if (!team.earendel.andromeda.configurations.GlobalConfiguration.get().paperPullRequestsPort.useVelocityOptimizeVarInts.enabled) return getVarIntSize_(value);
+        return VARINT_EXACT_BYTE_LENGTHS[Integer.numberOfLeadingZeros(value)];
+    }
+
+    private static final int[] VARINT_EXACT_BYTE_LENGTHS = new int[33];
+    static {
+        for (int i = 0; i <= 32; ++i)
+            VARINT_EXACT_BYTE_LENGTHS[i] = (int) Math.ceil((31d - (i - 1)) / 7d);
+        VARINT_EXACT_BYTE_LENGTHS[32] = 1;
+    }
+
+    public static int getVarIntSize_(int value) {
+        // Andromeda end
         for (int j = 1; j < 5; ++j) {
             if ((value & -1 << j * 7) == 0) {
                 return j;
@@ -568,6 +582,19 @@ public class FriendlyByteBuf extends ByteBuf {
     }
 
     public FriendlyByteBuf writeVarInt(int value) {
+        // Andromeda start - Port PaperMC/Paper#8414
+        if (!team.earendel.andromeda.configurations.GlobalConfiguration.get().paperPullRequestsPort.useVelocityOptimizeVarInts.enabled)
+            return writeVarInt_(value);
+        else if ((value & (0xFFFFFFFF << 7)) == 0)
+            writeByte(value);
+        else if ((value & (0xFFFFFFFF << 14)) == 0)
+            writeShort((value & 0x7F | 0x80) << 8 | (value >>> 7));
+        else
+            return writeVarInt_(value);
+        return this;
+    }
+    public FriendlyByteBuf writeVarInt_(int value) {
+        // Andromeda end
         while ((value & -128) != 0) {
             this.writeByte(value & 127 | 128);
             value >>>= 7;
diff --git a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
index 10537599d0c634c7d790b28d2c0d8906ea99aa4a..c65e9328c627d57e1da44e3ab1c03a4ee888753f 100644
--- a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
+++ b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
@@ -157,6 +157,13 @@ public class GlobalConfiguration extends ConfigurationPart {
 
         }
 
+        public UseVelocityOptimizeVarInts useVelocityOptimizeVarInts;
+        public class UseVelocityOptimizeVarInts extends ConfigurationPart {
+
+            public boolean enabled = true;
+
+        }
+
     }
 
 }
