From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sun, 8 Jan 2023 12:04:07 +0900
Subject: [PATCH] Port PaperMC/Paper#8418


diff --git a/src/main/java/net/minecraft/network/FriendlyByteBuf.java b/src/main/java/net/minecraft/network/FriendlyByteBuf.java
index 1b271896213f40b007a87ebbfe44e5c774db6994..d9ed6db426dd4613733463ffcf7dc900b7109896 100644
--- a/src/main/java/net/minecraft/network/FriendlyByteBuf.java
+++ b/src/main/java/net/minecraft/network/FriendlyByteBuf.java
@@ -96,6 +96,19 @@ public class FriendlyByteBuf extends ByteBuf {
     }
 
     public static int getVarIntSize(int value) {
+        // Andromeda start - Port PaperMC/Paper#8414
+        return VARINT_EXACT_BYTE_LENGTHS[Integer.numberOfLeadingZeros(value)];
+    }
+
+    private static final int[] VARINT_EXACT_BYTE_LENGTHS = new int[33];
+    static {
+        for (int i = 0; i <= 32; ++i)
+            VARINT_EXACT_BYTE_LENGTHS[i] = (int) Math.ceil((31d - (i - 1)) / 7d);
+        VARINT_EXACT_BYTE_LENGTHS[32] = 1;
+    }
+
+    public static int getVarIntSize_(int value) {
+        // Andromeda end
         for (int j = 1; j < 5; ++j) {
             if ((value & -1 << j * 7) == 0) {
                 return j;
@@ -568,6 +581,17 @@ public class FriendlyByteBuf extends ByteBuf {
     }
 
     public FriendlyByteBuf writeVarInt(int value) {
+        // Andromeda start - Port PaperMC/Paper#8414
+        if ((value & (0xFFFFFFFF << 7)) == 0)
+            writeByte(value);
+        else if ((value & (0xFFFFFFFF << 14)) == 0)
+            writeShort((value & 0x7F | 0x80) << 8 | (value >>> 7));
+        else
+            writeVarInt_(value);
+        return this;
+    }
+    public FriendlyByteBuf writeVarInt_(int value) {
+        // Andromeda end
         while ((value & -128) != 0) {
             this.writeByte(value & 127 | 128);
             value >>>= 7;
