From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Thu, 12 Jan 2023 15:42:46 +0900
Subject: [PATCH] Port PaperMC/Paper#8695


diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index ca2e0c1f95484e73383f6d5800c285f0b6e48f5b..0acb93af3d935c8032e334e4d091a76846a5751b 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -162,6 +162,8 @@ public class LevelConfigurations extends ConfigurationPart {
 
         public boolean checkCommandBlockPermissionInCanUseGameMasterBlocks = true;
 
+        public boolean avoidRespawnDesyncOfPlayerMetadata = true;
+
     }
 
 }
