From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Thu, 12 Jan 2023 16:42:31 +0900
Subject: [PATCH] fixup! Port PaperMC/Paper#8524


diff --git a/src/main/java/net/minecraft/world/item/ArmorItem.java b/src/main/java/net/minecraft/world/item/ArmorItem.java
index eab2828ef278ac6c53211d4fb59ba45b2eb4ff61..e90341a4f3bf6313441504adeccde3e3e3bf87df 100644
--- a/src/main/java/net/minecraft/world/item/ArmorItem.java
+++ b/src/main/java/net/minecraft/world/item/ArmorItem.java
@@ -37,7 +37,7 @@ public class ArmorItem extends Item implements Wearable {
     public static final DispenseItemBehavior DISPENSE_ITEM_BEHAVIOR = new DefaultDispenseItemBehavior() {
         @Override
         protected ItemStack execute(BlockSource pointer, ItemStack stack) {
-            return ArmorItem.dispenseArmor(pointer, stack) ? stack : super.execute(pointer, stack);
+            return ArmorItem.dispenseArmor(pointer, stack, null) ? stack : super.execute(pointer, stack);
         }
     };
     protected final EquipmentSlot slot;
