From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Thu, 12 Jan 2023 18:07:36 +0900
Subject: [PATCH] Port PaperMC/Paper#8075


diff --git a/src/main/java/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java b/src/main/java/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
index 84b230d979a91dd776e180e5b828b26bdc98ef12..77646b76acde31e12739970ed2ab601b3c833ea1 100644
--- a/src/main/java/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/src/main/java/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -617,6 +617,7 @@ public class EnderDragon extends Mob implements Enemy {
                 this.level.removeBlock(new BlockPos(block.getX(), block.getY(), block.getZ()), false);
             }
         } else {
+            if (this.level.andromedaLevelConfiguration().paperPullRequestsPort.doNotFireTntPrimeEventForEnderDragon) flag1 = false; // Andromeda - Port PaperMC/Paper#8075
             for (org.bukkit.block.Block block : event.blockList()) {
                 org.bukkit.Material blockId = block.getType();
                 if (blockId.isAir()) {
@@ -636,14 +637,18 @@ public class EnderDragon extends Mob implements Enemy {
                     });
                     craftBlock.getNMS().spawnAfterBreak((ServerLevel) level, blockposition, ItemStack.EMPTY, false);
                 }
+                if (!this.level.andromedaLevelConfiguration().paperPullRequestsPort.doNotFireTntPrimeEventForEnderDragon) { // Andromeda - Port PaperMC/Paper#8075
                 // Paper start - TNTPrimeEvent
                 org.bukkit.block.Block tntBlock = level.getWorld().getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ());
                 if(!new com.destroystokyo.paper.event.block.TNTPrimeEvent(tntBlock, com.destroystokyo.paper.event.block.TNTPrimeEvent.PrimeReason.EXPLOSION, explosionSource.getIndirectSourceEntity().getBukkitEntity()).callEvent())
                     continue;
                 // Paper end
                 nmsBlock.wasExploded(level, blockposition, explosionSource);
+                } // Andromeda - Port PaperMC/Paper#8075
 
+                boolean flag2 = // Andromeda - Port PaperMC/Paper#8075
                 this.level.removeBlock(blockposition, false);
+                if (this.level.andromedaLevelConfiguration().paperPullRequestsPort.doNotFireTntPrimeEventForEnderDragon) flag1 = flag1 || flag2; // Andromeda - Port PaperMC/Paper#8075
             }
         }
         // CraftBukkit end
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index c1956563cddda7230059234b2b906a434a5dc1cd..d0cb26fc251a09f3b98b913c755cd4067c5a2a51 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -168,6 +168,8 @@ public class LevelConfigurations extends ConfigurationPart {
 
         public boolean useSpigotConfigForMangrovePropaguleGrowth = true;
 
+        public boolean doNotFireTntPrimeEventForEnderDragon = true;
+
     }
 
 }
