From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Fri, 13 Jan 2023 15:22:52 +0900
Subject: [PATCH] Port PaperMC/Paper#7781


diff --git a/src/main/java/net/minecraft/world/level/chunk/UpgradeData.java b/src/main/java/net/minecraft/world/level/chunk/UpgradeData.java
index 804611147afa95b04059b375cbf81f32d63541eb..585c62b1b71cb9184cf9d3ac7ca8d4a893fa1571 100644
--- a/src/main/java/net/minecraft/world/level/chunk/UpgradeData.java
+++ b/src/main/java/net/minecraft/world/level/chunk/UpgradeData.java
@@ -357,7 +357,7 @@ public class UpgradeData {
         STEM_BLOCK(Blocks.MELON_STEM, Blocks.PUMPKIN_STEM) {
             @Override
             public BlockState updateShape(BlockState oldState, Direction direction, BlockState otherState, LevelAccessor world, BlockPos currentPos, BlockPos otherPos) {
-                if (oldState.getValue(StemBlock.AGE) == 7) {
+                if (oldState.getValue(StemBlock.AGE) == 7 && direction.getAxis().isHorizontal()) { // Andromeda - Port PaperMC/Paper#7781
                     StemGrownBlock stemGrownBlock = ((StemBlock)oldState.getBlock()).getFruit();
                     if (otherState.is(stemGrownBlock)) {
                         return stemGrownBlock.getAttachedStem().defaultBlockState().setValue(HorizontalDirectionalBlock.FACING, direction);
