From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Fri, 13 Jan 2023 18:36:11 +0900
Subject: [PATCH] Port PaperMC/Paper#5156


diff --git a/src/main/java/net/minecraft/world/entity/monster/EnderMan.java b/src/main/java/net/minecraft/world/entity/monster/EnderMan.java
index c029d6eb7893d41432e3de15fbf94768ef595d3f..fc0eebda32fe193c725252171c5277c59d52263d 100644
--- a/src/main/java/net/minecraft/world/entity/monster/EnderMan.java
+++ b/src/main/java/net/minecraft/world/entity/monster/EnderMan.java
@@ -133,8 +133,9 @@ public class EnderMan extends Monster implements NeutralMob {
         this.targetSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur
         this.targetSelector.addGoal(1, new EnderMan.EndermanLookForPlayerGoal(this, this::isAngryAt));
         this.targetSelector.addGoal(2, new HurtByTargetGoal(this, new Class[0]));
-        this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, Endermite.class, 10, true, false, (entityliving) -> entityliving.level.purpurConfig.endermanAggroEndermites && entityliving instanceof Endermite endermite && (!entityliving.level.purpurConfig.endermanAggroEndermitesOnlyIfPlayerSpawned || endermite.isPlayerSpawned()))); // Purpur
-        this.targetSelector.addGoal(4, new ResetUniversalAngerTargetGoal<>(this, false));
+        if (this.level.andromedaLevelConfiguration().entity.monster.enderMan.affectUniversalAnger) this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, Player.class, 10, true, false, this::isAngryAt));
+        this.targetSelector.addGoal(this.level.andromedaLevelConfiguration().entity.monster.enderMan.affectUniversalAnger ? 4 : 3 , new NearestAttackableTargetGoal<>(this, Endermite.class, 10, true, false, (entityliving) -> entityliving.level.purpurConfig.endermanAggroEndermites && entityliving instanceof Endermite endermite && (!entityliving.level.purpurConfig.endermanAggroEndermitesOnlyIfPlayerSpawned || endermite.isPlayerSpawned()))); // Purpur
+        this.targetSelector.addGoal(this.level.andromedaLevelConfiguration().entity.monster.enderMan.affectUniversalAnger ? 5 : 4, new ResetUniversalAngerTargetGoal<>(this, false));
     }
 
     public static AttributeSupplier.Builder createAttributes() {
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 6adead94485ad859f8566be56f0e5dc4fbe70ef7..81530fbe101b54f65afa303b781d3bb1169bcd7c 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -68,6 +68,17 @@ public class LevelConfigurations extends ConfigurationPart {
 
         }
 
+        public Monster monster;
+        public class Monster extends ConfigurationPart {
+
+            public EnderMan enderMan;
+
+            public class EnderMan extends ConfigurationPart {
+                public boolean affectUniversalAnger = false;
+            }
+
+        }
+
     }
 
     public Structure structure;
