From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Fri, 13 Jan 2023 20:31:40 +0900
Subject: [PATCH] Port PaperMC/Paper#7171


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 9895274b7a2c359907c94ac88631055cff809dd2..fd2b1969649e0698599fcfb72ab6e1e8ea8aeb9a 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -287,11 +287,26 @@ public class Explosion {
                             d14 = entity instanceof Player && level.paperConfig().environment.disableExplosionKnockback ? 0 : ProtectionEnchantment.getExplosionKnockbackAfterDampener((LivingEntity) entity, d13); // Paper - Disable explosion knockback
                         }
 
+                        // Andromeda start - Port PaperMC/Paper#7171
+                        org.bukkit.util.Vector delta = null;
+                        if (this.level.andromedaLevelConfiguration().paperPullRequestsPort.fireEntityKnockbackEventForOwnerlessTnt) {
+                            delta = new org.bukkit.util.Vector(d8 * d14, d9 * d14, d10 * d14);
+                            if (this.source != null && this.damageSource.getEntity() == null && entity instanceof LivingEntity livingEntity)
+                                if (!new com.destroystokyo.paper.event.entity.EntityKnockbackByEntityEvent(livingEntity.getBukkitLivingEntity(), this.source.getBukkitEntity(), (float) d14, delta).callEvent())
+                                    return;
+                            entity.setDeltaMovement(entity.getDeltaMovement().add(delta.getX(), delta.getY(), delta.getZ()));
+                        } else
+                        // Andromeda end
                         entity.setDeltaMovement(entity.getDeltaMovement().add(d8 * d14, d9 * d14, d10 * d14));
                         if (entity instanceof Player) {
                             Player entityhuman = (Player) entity;
 
                             if (!entityhuman.isSpectator() && (!entityhuman.isCreative() || !entityhuman.getAbilities().flying) && !level.paperConfig().environment.disableExplosionKnockback) { // Paper - Disable explosion knockback
+                                // Andromeda start - Port PaperMC/Paper#7171
+                                if (this.level.andromedaLevelConfiguration().paperPullRequestsPort.fireEntityKnockbackEventForOwnerlessTnt)
+                                    this.hitPlayers.put(entityhuman, org.bukkit.craftbukkit.util.CraftVector.toNMS(delta));
+                                else
+                                // Andromeda end
                                 this.hitPlayers.put(entityhuman, new Vec3(d8 * d13, d9 * d13, d10 * d13));
                             }
                         }
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 5ca9f2d0bda50190205918e00ad826922b7030e4..e3c6a7356d6e5ed8eef0ea5ab47931e2790d5e28 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -185,6 +185,8 @@ public class LevelConfigurations extends ConfigurationPart {
 
         public boolean fixHumanEntityDropNotUpdatingTheClientInventory = true;
 
+        public boolean fireEntityKnockbackEventForOwnerlessTnt = true;
+
     }
 
 }
