From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Fri, 13 Jan 2023 20:25:25 +0900
Subject: [PATCH] Port PaperMC/Paper#6761


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 724bcbadb1e96f251b412b0fa8aa6de5247c3ec6..44ad5dc9f580dd47915e4585c9f74e0a399fe561 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -228,7 +228,7 @@ public class ServerPlayer extends Player {
     private boolean textFilteringEnabled;
     private boolean allowsListing;
     public WardenSpawnTracker wardenSpawnTracker;
-    private final ContainerSynchronizer containerSynchronizer;
+    public final ContainerSynchronizer containerSynchronizer; // Andromeda - private -> public
     private final ContainerListener containerListener;
     @Nullable
     private RemoteChatSession chatSession;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 33cf1e70b22eca54a245987b12f37efaccf7d1e8..df40eab286e583644e74472e6aef1e40ecace8b4 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -750,8 +750,19 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
     // Paper end
     @Override
     public boolean dropItem(boolean dropAll) {
+        // Andromeda start - Port PaperMC/Paper#6761
+        if (this.getHandle().level.andromedaLevelConfiguration().paperPullRequestsPort.fixHumanEntityDropNotUpdatingTheClientInventory) {
+            if (!(this.getHandle() instanceof ServerPlayer player)) return false;
+            boolean success = player.drop(dropAll);
+            final net.minecraft.world.entity.player.Inventory inv = player.getInventory();
+            final java.util.OptionalInt optionalSlot = player.containerMenu.findSlot(inv, inv.selected);
+            optionalSlot.ifPresent(slot -> player.containerSynchronizer.sendSlotChange(player.containerMenu, slot, inv.getSelected()));
+            return success;
+        } else {
+        // Andromeda end
         if (!(this.getHandle() instanceof ServerPlayer)) return false;
         return ((ServerPlayer) this.getHandle()).drop(dropAll);
+        } // Andromeda - Port PaperMC/Paper#6761
     }
 
     @Override
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 12714c878d0c63c0aae35be1f6e46bbc416d6999..974579b5333d94fa18e3de179fe5874489e3da21 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -185,6 +185,8 @@ public class LevelConfigurations extends ConfigurationPart {
 
         public boolean improveEntityTransformEventCancellationHandling = true;
 
+        public boolean fixHumanEntityDropNotUpdatingTheClientInventory = true;
+
     }
 
 }
