From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sun, 15 Jan 2023 16:03:01 +0900
Subject: [PATCH] Mining speed tweaks


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index f23fdf5ff760eac45f0c35acf1c766e0323d7135..4db190ba8ac05786d4923afb84032ec1f176d864 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -185,6 +185,7 @@ public abstract class Player extends LivingEntity {
     // Paper end
     public int sixRowEnderchestSlotCount = -1; // Purpur
     public boolean canPortalInstant = false; // Purpur
+    public float miningSpeedAmplifier = 0; // Andromeda
 
     // CraftBukkit start
     public boolean fauxSleeping;
@@ -871,14 +872,16 @@ public abstract class Player extends LivingEntity {
             f *= f1;
         }
 
-        if (this.isEyeInFluid(FluidTags.WATER) && !EnchantmentHelper.hasAquaAffinity(this)) {
+        if (this.isEyeInFluid(FluidTags.WATER) && !EnchantmentHelper.hasAquaAffinity(this)  && !this.getLevel().andromedaLevelConfiguration().entity.player.mining.disableSpeedReductionWhenSwiming.enabled && (!this.getLevel().andromedaLevelConfiguration().entity.player.mining.disableSpeedReductionWhenSwiming.usePermission || this.getBukkitEntity().hasPermission("andromeda.mining.bypass-swiming-reduction"))) { // Andromeda
             f /= 5.0F;
         }
 
-        if (!this.onGround) {
+        if (!this.onGround && !this.getLevel().andromedaLevelConfiguration().entity.player.mining.disableSpeedReductionWhenFlying.enabled && (!this.getLevel().andromedaLevelConfiguration().entity.player.mining.disableSpeedReductionWhenFlying.usePermission || this.getBukkitEntity().hasPermission("andromeda.mining.bypass-flying-reduction"))) { // Andromeda
             f /= 5.0F;
         }
 
+        f += this.miningSpeedAmplifier; // Andromeda
+
         return f;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index e5dd91425159d28be40a155d5acb091955d7cf75..073564b3959d42817647c975491f3eab5e8be50e 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3179,5 +3179,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public void setModName(net.kyori.adventure.text.Component component) {
         this.getHandle().connection.send(new ClientboundCustomPayloadPacket(ClientboundCustomPayloadPacket.BRAND, (new FriendlyByteBuf(Unpooled.buffer())).writeUtf(io.papermc.paper.adventure.PaperAdventure.asJsonString(component, java.util.Locale.ROOT))));
     }
+
+    @Override
+    public float getMiningAmplifier() {
+        return this.getHandle().miningSpeedAmplifier;
+    }
+
+    @Override
+    public void setMiningAmplifier(float amplifier) {
+        this.getHandle().miningSpeedAmplifier = amplifier;
+    }
     // Andromeda end
 }
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 1bcd30e8d841a97bdb9d1138c7c07fc9201a1947..008f2af627250b0a3f420de13f26b6400a3c74f6 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -32,6 +32,27 @@ public class LevelConfigurations extends ConfigurationPart {
             public boolean checkSpectatorMovedToQuickly = true;
             public boolean giveAdvancementsReward = true;
 
+            public Mining mining;
+            public class Mining extends ConfigurationPart {
+
+                public DisableSpeedReductionWhenFlying disableSpeedReductionWhenFlying;
+                public class DisableSpeedReductionWhenFlying extends ConfigurationPart {
+
+                    public boolean enabled = false;
+                    public boolean usePermission = false;
+
+                }
+
+                public DisableSpeedReductionWhenSwiming disableSpeedReductionWhenSwiming;
+                public class DisableSpeedReductionWhenSwiming extends ConfigurationPart {
+
+                    public boolean enabled = false;
+                    public boolean usePermission = false;
+
+                }
+
+            }
+
         }
 
         public Vehicle vehicle;
