From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 14 Jan 2023 02:08:47 +0900
Subject: [PATCH] Port PaperMC/Paper#7603


diff --git a/src/main/java/io/papermc/paper/console/BrigadierCommandCompleter.java b/src/main/java/io/papermc/paper/console/BrigadierCommandCompleter.java
index 0627c98cae0b5ebdd71a849ae1299d7d3d581850..328d68608aae9e3828a5f37d69770d7082213547 100644
--- a/src/main/java/io/papermc/paper/console/BrigadierCommandCompleter.java
+++ b/src/main/java/io/papermc/paper/console/BrigadierCommandCompleter.java
@@ -39,7 +39,7 @@ public final class BrigadierCommandCompleter {
             return;
         }
         final CommandDispatcher<CommandSourceStack> dispatcher = this.server.getCommands().getDispatcher();
-        final ParseResults<CommandSourceStack> results = dispatcher.parse(prepareStringReader(line.line()), this.commandSourceStack.get());
+        final ParseResults<CommandSourceStack> results = dispatcher.parse(new StringReader(line.line()), this.commandSourceStack.get()); // Andromeda - Port PaperMC/Paper#7603
         this.addCandidates(
             candidates,
             dispatcher.getCompletionSuggestions(results, line.cursor()).join().getList(),
@@ -88,12 +88,4 @@ public final class BrigadierCommandCompleter {
         }
         return completion(suggestion.getText(), PaperAdventure.asAdventure(ComponentUtils.fromMessage(suggestion.getTooltip())));
     }
-
-    static @NonNull StringReader prepareStringReader(final @NonNull String line) {
-        final StringReader stringReader = new StringReader(line);
-        if (stringReader.canRead() && stringReader.peek() == '/') {
-            stringReader.skip();
-        }
-        return stringReader;
-    }
 }
diff --git a/src/main/java/io/papermc/paper/console/BrigadierCommandHighlighter.java b/src/main/java/io/papermc/paper/console/BrigadierCommandHighlighter.java
index dd9d77d7c7f1a5a130a1f4c15e5b1e68ae3753e1..17d28cf8d9fcd762c3c723264b8257a18929e615 100644
--- a/src/main/java/io/papermc/paper/console/BrigadierCommandHighlighter.java
+++ b/src/main/java/io/papermc/paper/console/BrigadierCommandHighlighter.java
@@ -2,6 +2,7 @@ package io.papermc.paper.console;
 
 import com.google.common.base.Suppliers;
 import com.mojang.brigadier.ParseResults;
+import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.context.ParsedCommandNode;
 import com.mojang.brigadier.tree.LiteralCommandNode;
 import java.util.function.Supplier;
@@ -32,12 +33,8 @@ public final class BrigadierCommandHighlighter implements Highlighter {
             return new AttributedString(buffer, AttributedStyle.DEFAULT.foreground(AttributedStyle.RED));
         }
         final AttributedStringBuilder builder = new AttributedStringBuilder();
-        final ParseResults<CommandSourceStack> results = this.server.getCommands().getDispatcher().parse(BrigadierCommandCompleter.prepareStringReader(buffer), this.commandSourceStack.get());
+        final ParseResults<CommandSourceStack> results = this.server.getCommands().getDispatcher().parse(new StringReader(buffer), this.commandSourceStack.get()); // Andromeda - Port PaperMC/Paper#7603
         int pos = 0;
-        if (buffer.startsWith("/")) {
-            builder.append("/", AttributedStyle.DEFAULT);
-            pos = 1;
-        }
         int component = -1;
         for (final ParsedCommandNode<CommandSourceStack> node : results.getContext().getLastChild().getNodes()) {
             if (node.getRange().getStart() >= buffer.length()) {
diff --git a/src/main/java/org/bukkit/craftbukkit/command/ConsoleCommandCompleter.java b/src/main/java/org/bukkit/craftbukkit/command/ConsoleCommandCompleter.java
index cd4ad8261e56365850068db1d83d6a8454026737..b5c2a9482a300f0d3852705cabbca7f137d1d0f7 100644
--- a/src/main/java/org/bukkit/craftbukkit/command/ConsoleCommandCompleter.java
+++ b/src/main/java/org/bukkit/craftbukkit/command/ConsoleCommandCompleter.java
@@ -29,7 +29,7 @@ public class ConsoleCommandCompleter implements Completer {
     @Override
     public void complete(LineReader reader, ParsedLine line, List<Candidate> candidates) {
         final CraftServer server = this.server.server;
-        final String buffer = line.line();
+        final String buffer = "/" + line.line(); // Andromeda - Port PaperMC/Paper#7603
         // Async Tab Complete
         final com.destroystokyo.paper.event.server.AsyncTabCompleteEvent event =
             new com.destroystokyo.paper.event.server.AsyncTabCompleteEvent(server.getConsoleSender(), buffer, true, null);
