From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 14 Jan 2023 02:29:04 +0900
Subject: [PATCH] Backport CraftBukkit@7dcb59b


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaMusicInstrument.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaMusicInstrument.java
index be793b06d0ba9e9cebb6c8af5b688ae1cfbfe2e9..d7e8d495c77c5b38215ae4c3eef2db8ce2f9eea5 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaMusicInstrument.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaMusicInstrument.java
@@ -53,12 +53,7 @@ public class CraftMetaMusicInstrument extends CraftMetaItem implements MusicInst
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case GOAT_HORN:
-                return true;
-            default:
-                return false;
-        }
+        return type == Material.GOAT_HORN; // Andromeda - backport CraftBukkit@7dcb59b
     }
 
     @Override
