From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Mon, 16 Jan 2023 14:30:01 +0900
Subject: [PATCH] Port PaperMC/Paper#8792


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 3e5abea2f814a0a364cf87ff4ce1b1718ba2ddac..67ae3270e477115984fd92c0e80975013f0c7ddd 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -214,6 +214,23 @@ public final class CraftItemStack extends ItemStack {
         if (this.handle == null) {
             return 0;
         }
+
+        // Andromeda start - Port PaperMC/Paper#8792
+        if (team.earendel.andromeda.configurations.GlobalConfiguration.get().paperPullRequestsPort.fixCisCimEnchantmentLevelInconsistency) {
+            ListTag enchantments = this.handle.getEnchantmentTags();
+
+            for (int i = 0; i < enchantments.size(); i++) {
+                CompoundTag tag = enchantments.getCompound(i);
+                String id = tag.getString(ENCHANTMENTS_ID.NBT);
+                Enchantment enchantment = Enchantment.getByKey(CraftNamespacedKey.fromStringOrNull(id));
+
+                if (ench.equals(enchantment))
+                    return 0xffff & tag.getShort(ENCHANTMENTS_LVL.NBT);
+            }
+
+            return 0;
+        } else
+        // Andromeda end
         return net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(CraftEnchantment.getRaw(ench), handle); // Paper
     }
 
diff --git a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
index e395280c3ff0849585ce0864f29b18cfaa516017..3f0ccbbe7e198a3f81047c085bd3f9fe9f35c780 100644
--- a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
+++ b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
@@ -147,6 +147,8 @@ public class GlobalConfiguration extends ConfigurationPart {
 
         public boolean properlyCloneCustomNbtTagsInsideItemMeta = true;
 
+        public boolean fixCisCimEnchantmentLevelInconsistency = true;
+
     }
 
 }
