From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@gmail.com>
Date: Mon, 6 Feb 2023 19:24:10 +0900
Subject: [PATCH] [Airplane] Use normal context whici is already lazy


diff --git a/src/main/java/io/papermc/paper/util/CollisionUtil.java b/src/main/java/io/papermc/paper/util/CollisionUtil.java
index a87f6380b2c387fb0cdd40d5087b5c93492e3c88..d827948a2c6d904c316c1e92de68bcb4916b7157 100644
--- a/src/main/java/io/papermc/paper/util/CollisionUtil.java
+++ b/src/main/java/io/papermc/paper/util/CollisionUtil.java
@@ -619,6 +619,7 @@ public final class CollisionUtil {
                                 if ((edgeCount != 1 || blockData.shapeExceedsCube()) && (edgeCount != 2 || blockData.getBlock() == Blocks.MOVING_PISTON)) {
                                     mutablePos.set(blockX, blockY, blockZ);
                                     if (collisionShape == null) {
+                                        if (team.earendel.andromeda.configurations.GlobalConfiguration.get().misc.useNormalCollisionContextWhichIsAlreadyLazy) collisionShape = CollisionContext.of(entity); else // Andromeda
                                         collisionShape = new LazyEntityCollisionContext(entity);
                                     }
                                     VoxelShape voxelshape2 = blockData.getCollisionShape(getter, mutablePos, collisionShape);
diff --git a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
index 3f0ccbbe7e198a3f81047c085bd3f9fe9f35c780..a3641b97abd2cd07fd3cc7bea639309172a06905 100644
--- a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
+++ b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
@@ -105,6 +105,7 @@ public class GlobalConfiguration extends ConfigurationPart {
     public class Misc extends ConfigurationPart {
 
         public boolean useFasterRandom = true;
+        public boolean useNormalCollisionContextWhichIsAlreadyLazy = false;
 
         public UnsupportedSettings unsupportedSettings;
         public class UnsupportedSettings extends ConfigurationPart {
