From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@gmail.com>
Date: Mon, 6 Feb 2023 19:46:39 +0900
Subject: [PATCH] AFK Distance


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 44ad5dc9f580dd47915e4585c9f74e0a399fe561..5f0cdde72604625001c0cd64350fc0fd0ab08ee6 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2084,6 +2084,14 @@ public class ServerPlayer extends Player {
     public void resetLastActionTime() {
         this.lastActionTime = Util.getMillis();
         this.setAfk(false); // Purpur
+        // Andromeda start
+        if (this.level.andromedaLevelConfiguration().world.gameMechanics.idleDistance.enabled) {
+            try {
+                this.getBukkitEntity().setViewDistance(this.getBukkitEntity().getWorld().getViewDistance());
+                this.getBukkitEntity().setSimulationDistance(this.getBukkitEntity().getWorld().getSimulationDistance());
+            } catch (IllegalStateException ignored) {}
+        }
+        // Andromeda end
     }
 
     // Purpur Start
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index e324c2f69b2a17092a84eebbbc51375731222e58..286fff614bcbb4e33715a528b4c4d31d33303d76 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -461,6 +461,12 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
             // Purpur start
             this.player.setAfk(true);
             if (!this.player.level.purpurConfig.idleTimeoutKick || kickPermissionCache.getUnchecked(this.player.getBukkitEntity())) {
+                if (this.player.level.andromedaLevelConfiguration().world.gameMechanics.idleDistance.enabled) {
+                    try {
+                        this.getCraftPlayer().setViewDistance(this.player.level.andromedaLevelConfiguration().world.gameMechanics.idleDistance.viewDistance);
+                        this.getCraftPlayer().setSimulationDistance(this.player.level.andromedaLevelConfiguration().world.gameMechanics.idleDistance.simulationDistance);
+                    } catch (IllegalStateException ignored) {}
+                }
                 return;
             }
             // Purpur end
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index b62b37ac95a15f9107154832f1d9010549f150b0..7a1e03cf5fe6ea5cd828a02fe495d5f3d175c087 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -130,6 +130,13 @@ public class LevelConfigurations extends ConfigurationPart {
 
             public boolean unlockSnowLayerLimit = false;
 
+            public IdleDistance idleDistance;
+            public class IdleDistance extends ConfigurationPart {
+                public boolean enabled = false;
+                public int viewDistance = 7;
+                public int simulationDistance = 4;
+            }
+
         }
 
     }
