From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@gmail.com>
Date: Mon, 6 Feb 2023 20:54:28 +0900
Subject: [PATCH] Frogs Eat Everything! Oh my!


diff --git a/src/main/java/net/minecraft/world/entity/animal/frog/Frog.java b/src/main/java/net/minecraft/world/entity/animal/frog/Frog.java
index 71cb2ccc0b473ceccb25f25ee87be846a48c7781..1081fc099be5e5b5a120ba04b9bc0e3472a5da60 100644
--- a/src/main/java/net/minecraft/world/entity/animal/frog/Frog.java
+++ b/src/main/java/net/minecraft/world/entity/animal/frog/Frog.java
@@ -430,6 +430,7 @@ public class Frog extends Animal implements VariantHolder<FrogVariant> {
     }
 
     public static boolean canEat(LivingEntity entity) {
+        if (entity.level.andromedaLevelConfiguration().entity.animal.frog.eatsEverything) return true; // Andromeda
         if (entity instanceof Slime slime) {
             if (slime.getSize() != 1) {
                 return false;
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index b4d339a9c2088de96e888d4880f25d1bb559275f..36e8c40e46863858edc5c5738dfdec72ff6c2e53 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -126,6 +126,18 @@ public class LevelConfigurations extends ConfigurationPart {
 
         }
 
+        public Animal animal;
+        public class Animal extends ConfigurationPart {
+
+            public Frog frog;
+            public class Frog extends ConfigurationPart {
+
+                public boolean eatsEverything = false;
+
+            }
+
+        }
+
     }
 
     public Structure structure;
