From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@gmail.com>
Date: Fri, 10 Feb 2023 00:04:38 +0900
Subject: [PATCH] Port PaperMC/Paper#8834


diff --git a/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java b/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
index 0b060183429f4c72ec767075538477b4302bbf0d..0a796fad50467114d273e06f65a58a57de61e99d 100644
--- a/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
+++ b/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
@@ -383,15 +383,15 @@ public final class PlayerChunkLoader {
     protected int getMaxChunkLoads() {
         double config = GlobalConfiguration.get().chunkLoading.playerMaxConcurrentLoads;
         double max = GlobalConfiguration.get().chunkLoading.globalMaxConcurrentLoads;
-        return (int)Math.ceil(Math.min(config * MinecraftServer.getServer().getPlayerCount(), max <= 1.0 ? Double.MAX_VALUE : max));
+        return (int)Math.ceil(Math.min(config * MinecraftServer.getServer().getPlayerCount(), (team.earendel.andromeda.configurations.GlobalConfiguration.get().paperPullRequestsPort.fixChunkLoadingWhenTheRateIsSetToOne ? max < 1.0 : max <= 1.0) ? Double.MAX_VALUE : max)); // Andromeda
     }
 
     protected long getTargetSendPerPlayerAddend() {
-        return GlobalConfiguration.get().chunkLoading.targetPlayerChunkSendRate <= 1.0 ? 0L : (long)Math.round(1.0e9 / GlobalConfiguration.get().chunkLoading.targetPlayerChunkSendRate);
+        return (team.earendel.andromeda.configurations.GlobalConfiguration.get().paperPullRequestsPort.fixChunkLoadingWhenTheRateIsSetToOne ? GlobalConfiguration.get().chunkLoading.targetPlayerChunkSendRate < 1.0e-3 : GlobalConfiguration.get().chunkLoading.targetPlayerChunkSendRate <= 1.0) ? 0L : (long)Math.round(1.0e9 / GlobalConfiguration.get().chunkLoading.targetPlayerChunkSendRate); // Andromeda
     }
 
     protected long getMaxSendAddend() {
-        return GlobalConfiguration.get().chunkLoading.globalMaxChunkSendRate <= 1.0 ? 0L : (long)Math.round(1.0e9 / GlobalConfiguration.get().chunkLoading.globalMaxChunkSendRate);
+        return (team.earendel.andromeda.configurations.GlobalConfiguration.get().paperPullRequestsPort.fixChunkLoadingWhenTheRateIsSetToOne ? GlobalConfiguration.get().chunkLoading.globalMaxChunkSendRate < 1.0e-3 : GlobalConfiguration.get().chunkLoading.globalMaxChunkSendRate <= 1.0) ? 0L : (long)Math.round(1.0e9 / GlobalConfiguration.get().chunkLoading.globalMaxChunkSendRate); // Andromeda
     }
 
     public void onChunkPlayerTickReady(final int chunkX, final int chunkZ) {
diff --git a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
index 9f0dcd0fc57f8361a9b6ae101072a5e8cdff10a8..cb3bd4d1578332b02c816ad76a76f8d2720b4941 100644
--- a/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
+++ b/src/main/java/team/earendel/andromeda/configurations/GlobalConfiguration.java
@@ -171,6 +171,8 @@ public class GlobalConfiguration extends ConfigurationPart {
 
         public boolean fixCisCimEnchantmentLevelInconsistency = true;
 
+        public boolean fixChunkLoadingWhenTheRateIsSetToOne = true;
+
     }
 
 }
